From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sun, 3 Dec 2023 14:20:30 +0100
Subject: [PATCH] Implementing Basic Item Format


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 4697df75fdee2023c41260bed211e3e3d90d2b9b..40bfcb1fe51a8713a8d2c0ffa4bb1291ffd1afca 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -223,6 +223,7 @@ public final class ItemStack {
     private ItemStack(ItemLike item, int count, Optional<CompoundTag> nbt) {
         this(item, count);
         nbt.ifPresent(this::setTag);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     public ItemStack(Holder<Item> itemEntry, int count) {
@@ -235,7 +236,7 @@ public final class ItemStack {
         if (this.item.canBeDepleted()) {
             this.setDamageValue(this.getDamageValue());
         }
-
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     private ItemStack(@Nullable Void ovoid) {
@@ -273,6 +274,7 @@ public final class ItemStack {
 
     private ItemStack(CompoundTag nbt) {
         this.load(nbt);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
         // CraftBukkit end
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 6556d7ab09826bb5a99f11385eddc26b67e44d68..47363351557d0bc76d2806d3edee25f267d01fc0 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -120,6 +120,7 @@ public final class CraftItemStack extends ItemStack {
         this.setAmount(amount);
         this.setDurability(durability);
         this.setItemMeta(itemMeta);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(this);
     }
 
     @Override
