From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Fri, 17 Nov 2023 17:05:03 +0100
Subject: [PATCH] Adding isRainingAt method to World.java


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 19912dbe461eec708c2c2a025bf295de093e13b0..6621cb051be5a984b6430e67ec8e2f121571c9da 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -157,6 +157,10 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     // MCCLab - start
     @Override
+    public boolean isRainingAt(Location location) {
+        return world.isRainingAt(new BlockPos(location.getBlockX(),location.getBlockY(),location.getBlockZ()));
+    }
+    @Override
     public boolean growTreeIfSapling(@NotNull Location location) {
         var blockPos = new BlockPos(location.getBlockX(),location.getBlockY(),location.getBlockZ());
         var block = world.getBlockIfLoaded(blockPos);
