--- a/net/minecraft/util/BlockUtil.java
+++ b/net/minecraft/util/BlockUtil.java
@@ -16,6 +_,7 @@
     public static BlockUtil.FoundRectangle getLargestRectangleAround(
         BlockPos centerPos, Direction.Axis axis1, int max1, Direction.Axis axis2, int max2, Predicate<BlockPos> posPredicate
     ) {
+        System.out.println("hello");
         BlockPos.MutableBlockPos mutableBlockPos = centerPos.mutable();
         Direction direction = Direction.get(Direction.AxisDirection.NEGATIVE, axis1);
         Direction opposite = direction.getOpposite();
