From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Wed, 1 May 2024 15:04:15 +0000
Subject: [PATCH] Make ChunkDataEvents use WorldGenChunk


diff --git a/src/main/java/de/verdox/mccreativelab/events/ChunkDataCreateEvent.java b/src/main/java/de/verdox/mccreativelab/events/ChunkDataCreateEvent.java
index 578988593db6689ef98d53da2accc553a720d3b1..8086bd8e34cbca1c46f955dfe02590578bd2b42a 100644
--- a/src/main/java/de/verdox/mccreativelab/events/ChunkDataCreateEvent.java
+++ b/src/main/java/de/verdox/mccreativelab/events/ChunkDataCreateEvent.java
@@ -1,5 +1,6 @@
 package de.verdox.mccreativelab.events;
 
+import de.verdox.mccreativelab.worldgen.WorldGenChunk;
 import org.bukkit.Chunk;
 import org.bukkit.World;
 import org.bukkit.persistence.PersistentDataContainer;
@@ -7,15 +8,7 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 public class ChunkDataCreateEvent extends ChunkDataEvent {
-    private final Chunk chunk;
-
-    public ChunkDataCreateEvent(@NotNull World world, @Nullable Chunk chunk, @NotNull ChunkDataEvent.ChunkPos chunkPos, @NotNull PersistentDataContainer persistentDataContainer) {
+    public ChunkDataCreateEvent(@NotNull World world, WorldGenChunk chunk, @NotNull ChunkDataEvent.ChunkPos chunkPos, @NotNull PersistentDataContainer persistentDataContainer) {
         super(world, chunk, chunkPos, persistentDataContainer);
-        this.chunk = chunk;
-    }
-
-    @NotNull
-    public Chunk getChunk() {
-        return this.chunk;
     }
 }
diff --git a/src/main/java/de/verdox/mccreativelab/events/ChunkDataEvent.java b/src/main/java/de/verdox/mccreativelab/events/ChunkDataEvent.java
index 11ade060418a9997fc9376a5518ad38628d4fd9d..58ae701930935619b001046f4d3899e6709fc82c 100644
--- a/src/main/java/de/verdox/mccreativelab/events/ChunkDataEvent.java
+++ b/src/main/java/de/verdox/mccreativelab/events/ChunkDataEvent.java
@@ -1,5 +1,6 @@
 package de.verdox.mccreativelab.events;
 
+import de.verdox.mccreativelab.worldgen.WorldGenChunk;
 import org.bukkit.Bukkit;
 import org.bukkit.Chunk;
 import org.bukkit.World;
@@ -14,13 +15,13 @@ public abstract class ChunkDataEvent extends Event {
 
     private final World world;
 
-    private final Chunk chunk;
+    private final WorldGenChunk chunk;
 
     private final ChunkPos chunkPos;
 
     private final PersistentDataContainer persistentDataContainer;
 
-    public ChunkDataEvent(@NotNull World world, @Nullable Chunk chunk, @NotNull ChunkPos chunkPos, @NotNull PersistentDataContainer persistentDataContainer) {
+    public ChunkDataEvent(@NotNull World world, WorldGenChunk chunk, @NotNull ChunkPos chunkPos, @NotNull PersistentDataContainer persistentDataContainer) {
         super(!Bukkit.isPrimaryThread());
         this.world = world;
         this.chunk = chunk;
@@ -34,7 +35,7 @@ public abstract class ChunkDataEvent extends Event {
     }
 
     @Nullable
-    public Chunk getChunk() {
+    public WorldGenChunk getChunk() {
         return this.chunk;
     }
 
diff --git a/src/main/java/de/verdox/mccreativelab/events/ChunkDataLoadEvent.java b/src/main/java/de/verdox/mccreativelab/events/ChunkDataLoadEvent.java
index 763d32aba1a36e0cc839b9757d54acc185644348..9ebab7ef7a53bce9f3fce3bd9d027fbea86d9dc8 100644
--- a/src/main/java/de/verdox/mccreativelab/events/ChunkDataLoadEvent.java
+++ b/src/main/java/de/verdox/mccreativelab/events/ChunkDataLoadEvent.java
@@ -1,5 +1,6 @@
 package de.verdox.mccreativelab.events;
 
+import de.verdox.mccreativelab.worldgen.WorldGenChunk;
 import org.bukkit.Chunk;
 import org.bukkit.World;
 import org.bukkit.persistence.PersistentDataContainer;
@@ -7,7 +8,7 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 public class ChunkDataLoadEvent extends ChunkDataEvent {
-    public ChunkDataLoadEvent(@NotNull World world, @Nullable Chunk chunk, @NotNull ChunkDataEvent.ChunkPos chunkPos, @NotNull PersistentDataContainer persistentDataContainer) {
+    public ChunkDataLoadEvent(@NotNull World world, WorldGenChunk chunk, @NotNull ChunkDataEvent.ChunkPos chunkPos, @NotNull PersistentDataContainer persistentDataContainer) {
         super(world, chunk, chunkPos, persistentDataContainer);
     }
 }
diff --git a/src/main/java/de/verdox/mccreativelab/events/ChunkDataSaveEvent.java b/src/main/java/de/verdox/mccreativelab/events/ChunkDataSaveEvent.java
index 65adb34f43b7e77b87fe146ab0ec7bae9132beff..de45a3f5c1eb9a00931d45dd7ed3f31f628dbf60 100644
--- a/src/main/java/de/verdox/mccreativelab/events/ChunkDataSaveEvent.java
+++ b/src/main/java/de/verdox/mccreativelab/events/ChunkDataSaveEvent.java
@@ -1,5 +1,6 @@
 package de.verdox.mccreativelab.events;
 
+import de.verdox.mccreativelab.worldgen.WorldGenChunk;
 import org.bukkit.Chunk;
 import org.bukkit.World;
 import org.bukkit.persistence.PersistentDataContainer;
@@ -9,7 +10,7 @@ import org.jetbrains.annotations.Nullable;
 public class ChunkDataSaveEvent extends ChunkDataEvent {
     private final boolean unloaded;
 
-    public ChunkDataSaveEvent(@NotNull World world, @Nullable Chunk chunk, @NotNull ChunkDataEvent.ChunkPos chunkPos, @NotNull PersistentDataContainer persistentDataContainer, boolean unloaded) {
+    public ChunkDataSaveEvent(@NotNull World world, WorldGenChunk chunk, @NotNull ChunkDataEvent.ChunkPos chunkPos, @NotNull PersistentDataContainer persistentDataContainer, boolean unloaded) {
         super(world, chunk, chunkPos, persistentDataContainer);
         this.unloaded = unloaded;
     }
