From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Sat, 18 May 2024 23:18:52 +0000
Subject: [PATCH] Adding the ability to change the server resourcepack


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 19d5a2910a629d6dccba1a966b55fc2a953a6da0..3f68b5f3d21779fda29214e44d5a1aa0f3a68d08 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -349,6 +349,16 @@ public final class Bukkit {
         return server.getServerResourcePack();
     }
 
+    // Paper start - Make server resourcepack changeable
+    /**
+     * Sets the server resource pack
+     * @param resourcePack the resource pack
+     */
+    public static void setServerResourcePack(@Nullable ResourcePack resourcePack){
+        server.setServerResourcePack(resourcePack);
+    }
+    // Paper end - Make server resourcepack changeable
+
     /**
      * Get the ServerTick Manager.
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 3be7c91f81d5f887ce736697b3dd151cc48e1f08..1918586fbd202ad561e90a79c87ca408c2447a2d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -289,6 +289,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @Nullable
     public ResourcePack getServerResourcePack();
 
+    // Paper start - Make server resourcepack changeable
+    /**
+     * Sets the server resource pack
+     * @param resourcePack the resource pack
+     */
+    public void setServerResourcePack(@Nullable ResourcePack resourcePack);
+    // Paper end - Make server resourcepack changeable
+
     /**
      * Gets the server resource pack uri, or empty string if not specified.
      *
diff --git a/src/main/java/org/bukkit/packs/ResourcePack.java b/src/main/java/org/bukkit/packs/ResourcePack.java
index ea921799ab464fb4555a2e90513635db19116b17..1cc55bbb58e990eb0f63c83a9e8d73ad967dc101 100644
--- a/src/main/java/org/bukkit/packs/ResourcePack.java
+++ b/src/main/java/org/bukkit/packs/ResourcePack.java
@@ -43,7 +43,7 @@ public interface ResourcePack {
      * @return the prompt
      */
     @Nullable
-    public String getPrompt();
+    public net.kyori.adventure.text.Component getPrompt(); // Paper - Make server resourcepack changeable
 
     /**
      * Gets if the resource pack is required by the server.
diff --git a/src/test/java/org/bukkit/plugin/TimedRegisteredListenerTest.java b/src/test/java/org/bukkit/plugin/TimedRegisteredListenerTest.java
index 9ed416ed57676c845833736f93ed6088513c6da4..90ee4cdc682434aed66a98a687b2c549a6f381c4 100644
--- a/src/test/java/org/bukkit/plugin/TimedRegisteredListenerTest.java
+++ b/src/test/java/org/bukkit/plugin/TimedRegisteredListenerTest.java
@@ -25,7 +25,7 @@ public class TimedRegisteredListenerTest {
 
         PlayerInteractEvent interactEvent = new PlayerInteractEvent(null, null, null, null, null);
         PlayerMoveEvent moveEvent = new PlayerMoveEvent(null, null, null);
-        BlockBreakEvent breakEvent = new BlockBreakEvent(null, null);
+        BlockBreakEvent breakEvent = new BlockBreakEvent(null, null, null);
 
         TimedRegisteredListener trl = new TimedRegisteredListener(listener, executor, EventPriority.NORMAL, plugin, false);
 
