From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Mon, 27 Nov 2023 19:21:13 +0100
Subject: [PATCH] Additional imports for CraftDataFactory


diff --git a/src/main/java/org/bukkit/craftbukkit/datafactory/CraftDataFactory.java b/src/main/java/org/bukkit/craftbukkit/datafactory/CraftDataFactory.java
index ad3561efcb7f2ca0c36bc1b89facd1b6273709ed..0fb6960738089c26d1654c6cd960d11ec5ce4249 100644
--- a/src/main/java/org/bukkit/craftbukkit/datafactory/CraftDataFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/datafactory/CraftDataFactory.java
@@ -1,9 +1,11 @@
 package org.bukkit.craftbukkit.datafactory;
 
 import org.bukkit.craftbukkit.datafactory.builder.CraftAttributeBuilder;
+import org.bukkit.craftbukkit.datafactory.builder.CraftPoiTypeBuilder;
 import org.bukkit.craftbukkit.datafactory.builder.CraftVillagerProfessionBuilder;
 import org.bukkit.datafactory.BukkitDataFactory;
 import org.bukkit.datafactory.builder.AttributeBuilder;
+import org.bukkit.datafactory.builder.PoiTypeBuilder;
 import org.bukkit.datafactory.builder.VillagerProfessionBuilder;
 
 public class CraftDataFactory implements BukkitDataFactory {
