From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lukas <lukasjonsson@outlook.de>
Date: Sun, 19 Nov 2023 14:32:31 +0100
Subject: [PATCH] Opening random tick behaviour to API


diff --git a/src/main/java/de/verdox/mccreativelab/behaviour/BehaviourUtil.java b/src/main/java/de/verdox/mccreativelab/behaviour/BehaviourUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..c37dd2fd5f2cd41f6d5ec9244918096b54bd6cd7
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/behaviour/BehaviourUtil.java
@@ -0,0 +1,46 @@
+package de.verdox.mccreativelab.behaviour;
+
+import de.verdox.mccreativelab.random.CraftVanillaRandomSource;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.BlockPos;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import java.util.Locale;
+
+public class BehaviourUtil {
+
+    public static boolean isRandomlyTicking(BlockState blockState){
+        Material material = getMaterial(blockState);
+        if(RandomTickBehaviour.RANDOM_TICK_BEHAVIOUR.isImplemented(material))
+            return RandomTickBehaviour.RANDOM_TICK_BEHAVIOUR.getBehaviour(material).isRandomlyTicking(blockState.createCraftBlockData());
+        return false;
+    }
+
+    public static boolean isRandomTickBehaviourImplemented(BlockState blockState){
+        Material material = getMaterial(blockState);
+        return RandomTickBehaviour.RANDOM_TICK_BEHAVIOUR.isImplemented(material);
+    }
+
+    public static boolean randomTick(BlockState blockState, ServerLevel world, BlockPos pos, RandomSource random){
+        Material material = getMaterial(blockState);
+        if(RandomTickBehaviour.RANDOM_TICK_BEHAVIOUR.isImplemented(material)) {
+            Block block = world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+            RandomTickBehaviour.RANDOM_TICK_BEHAVIOUR.getBehaviour(material).randomTick(block, new CraftVanillaRandomSource(random));
+            return true;
+        }
+        return false;
+    }
+
+    public static Material getMaterial(BlockState blockState){
+        Material material;
+        if(MinecraftServer.getServer() != null && MinecraftServer.getServer().isReady())
+            material = blockState.getBukkitMaterial();
+        else
+            material = Material.getMaterial(BuiltInRegistries.BLOCK.getKey(blockState.getBlock()).getPath().toUpperCase(Locale.ROOT));
+        return material;
+    }
+}
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index dbccbcb9b44e4efacdf53c2d161115cc20b36cff..64c6fcb01739f59d79f828969f9083a593c5767a 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -272,6 +272,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         // this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage)); // Spigot - moved up
         server.loadPlugins();
         server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
+        net.minecraft.world.level.block.Blocks.rebuildCache(); // MCCLab
         // CraftBukkit end
 
         // Paper start
diff --git a/src/main/java/net/minecraft/world/level/block/BambooSaplingBlock.java b/src/main/java/net/minecraft/world/level/block/BambooSaplingBlock.java
index fd0143944e4fecaf6f1ffa4561702178431575e9..c3dbda01cc30770df44f755890e8541c6bf25af5 100644
--- a/src/main/java/net/minecraft/world/level/block/BambooSaplingBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BambooSaplingBlock.java
@@ -38,6 +38,10 @@ public class BambooSaplingBlock extends Block implements BonemealableBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (random.nextFloat() < (world.spigotConfig.bambooModifier / (100.0f * 3)) && world.isEmptyBlock(pos.above()) && world.getRawBrightness(pos.above(), 0) >= 9) { // Spigot - SPIGOT-7159: Better modifier resolution
             this.growBamboo(world, pos);
         }
diff --git a/src/main/java/net/minecraft/world/level/block/BambooStalkBlock.java b/src/main/java/net/minecraft/world/level/block/BambooStalkBlock.java
index 05a508cb457b67078ccb08bd2228a59e73701cf3..cd8a5dca38d3e76f56254b82b24c2c138947ed17 100644
--- a/src/main/java/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -121,11 +121,19 @@ public class BambooStalkBlock extends Block implements BonemealableBlock {
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return (Integer) state.getValue(BambooStalkBlock.STAGE) == 0;
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if ((Integer) state.getValue(BambooStalkBlock.STAGE) == 0) {
             if (random.nextFloat() < (world.spigotConfig.bambooModifier / (100.0f * 3)) && world.isEmptyBlock(pos.above()) && world.getRawBrightness(pos.above(), 0) >= 9) { // Spigot - SPIGOT-7159: Better modifier resolution
                 int i = this.getHeightBelowUpToMax(world, pos) + 1;
diff --git a/src/main/java/net/minecraft/world/level/block/BeetrootBlock.java b/src/main/java/net/minecraft/world/level/block/BeetrootBlock.java
index 2bd714466fb86d4013bd3bceddecba19ce98df08..5c45040390bc89730aa239c5d3d218dc76c97ff6 100644
--- a/src/main/java/net/minecraft/world/level/block/BeetrootBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BeetrootBlock.java
@@ -41,6 +41,10 @@ public class BeetrootBlock extends CropBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (random.nextInt(3) != 0) {
             super.randomTick(state, world, pos, random);
         }
diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index d4cbff18adb62073a1dceb189043789620af6877..392ab770b004a673445f8596a7981e0eb63e6590 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -223,6 +223,10 @@ public class Block extends BlockBehaviour implements ItemLike {
     }
 
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return this.isRandomlyTicking;
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/BuddingAmethystBlock.java b/src/main/java/net/minecraft/world/level/block/BuddingAmethystBlock.java
index 2548244c089dff583c8c424dcaa5504caedf3f10..c1fb487cc0b1aa3cd852358afce590666d2d3e14 100644
--- a/src/main/java/net/minecraft/world/level/block/BuddingAmethystBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/BuddingAmethystBlock.java
@@ -19,6 +19,10 @@ public class BuddingAmethystBlock extends AmethystBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (random.nextInt(5) == 0) {
             Direction enumdirection = BuddingAmethystBlock.DIRECTIONS[random.nextInt(BuddingAmethystBlock.DIRECTIONS.length)];
             BlockPos blockposition1 = pos.relative(enumdirection);
diff --git a/src/main/java/net/minecraft/world/level/block/CactusBlock.java b/src/main/java/net/minecraft/world/level/block/CactusBlock.java
index 0003fb51ae3a6575575e10b4c86719f3061e2577..2230e80ecb7de233d6170c47519570a9af4820c2 100644
--- a/src/main/java/net/minecraft/world/level/block/CactusBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CactusBlock.java
@@ -45,6 +45,10 @@ public class CactusBlock extends Block {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         BlockPos blockposition1 = pos.above();
 
         if (world.isEmptyBlock(blockposition1)) {
diff --git a/src/main/java/net/minecraft/world/level/block/ChorusFlowerBlock.java b/src/main/java/net/minecraft/world/level/block/ChorusFlowerBlock.java
index e9beebedf1d60e06e16c442b64d28d022a3e3164..c08cf5dd5b5aee516ca482031e617c512bdecfe9 100644
--- a/src/main/java/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -45,6 +45,10 @@ public class ChorusFlowerBlock extends Block {
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return (Integer) state.getValue(ChorusFlowerBlock.AGE) < 5;
     }
 
@@ -55,6 +59,10 @@ public class ChorusFlowerBlock extends Block {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         BlockPos blockposition1 = pos.above();
 
         if (world.isEmptyBlock(blockposition1) && blockposition1.getY() < world.getMaxBuildHeight()) {
diff --git a/src/main/java/net/minecraft/world/level/block/CocoaBlock.java b/src/main/java/net/minecraft/world/level/block/CocoaBlock.java
index f5c42fee3ec68a7a9260f01d19c7a1d7533bad9c..f049edccafa877befb1790354a094579a0a72f4a 100644
--- a/src/main/java/net/minecraft/world/level/block/CocoaBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CocoaBlock.java
@@ -46,11 +46,19 @@ public class CocoaBlock extends HorizontalDirectionalBlock implements Bonemealab
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return (Integer) state.getValue(CocoaBlock.AGE) < 2;
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (world.random.nextFloat() < (world.spigotConfig.cocoaModifier / (100.0f * 5))) { // Spigot - SPIGOT-7159: Better modifier resolution
             int i = (Integer) state.getValue(CocoaBlock.AGE);
 
diff --git a/src/main/java/net/minecraft/world/level/block/CropBlock.java b/src/main/java/net/minecraft/world/level/block/CropBlock.java
index 6365ddea0c23bc5d4009d98915f2b39aed2a0328..844722ac6b069f6a4497b3c1f0aa622dbdb365aa 100644
--- a/src/main/java/net/minecraft/world/level/block/CropBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CropBlock.java
@@ -65,11 +65,19 @@ public class CropBlock extends BushBlock implements BonemealableBlock {
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return !this.isMaxAge(state);
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (world.getRawBrightness(pos, 0) >= 9) {
             int i = this.getAge(state);
 
diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index 5946f06f63b5694034bd027984a4925b0831d439..8e5a003ba6ab948aba4de7dc8f2a4b9adbdd8e60 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -84,6 +84,10 @@ public class FarmBlock extends Block {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         int i = (Integer) state.getValue(FarmBlock.MOISTURE);
 
         if (!FarmBlock.isNearWater(world, pos) && !world.isRainingAt(pos.above())) {
diff --git a/src/main/java/net/minecraft/world/level/block/FrostedIceBlock.java b/src/main/java/net/minecraft/world/level/block/FrostedIceBlock.java
index a3af96b2518b41f370d09cfda26dc589b9ee977b..595d8a53509fe678a0d8c00e98918518aa30e5fe 100644
--- a/src/main/java/net/minecraft/world/level/block/FrostedIceBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FrostedIceBlock.java
@@ -27,6 +27,10 @@ public class FrostedIceBlock extends IceBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         this.tick(state, world, pos, random);
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/GrowingPlantHeadBlock.java b/src/main/java/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
index b325fdec8eb99ce57152a5c0fefa5059685276e4..dc6314518127c39ddd4efbfa4e8b035306872c7c 100644
--- a/src/main/java/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -35,11 +35,19 @@ public abstract class GrowingPlantHeadBlock extends GrowingPlantBlock implements
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return (Integer) state.getValue(GrowingPlantHeadBlock.AGE) < 25;
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         // Spigot start
         int modifier;
         if (this == Blocks.KELP) {
diff --git a/src/main/java/net/minecraft/world/level/block/IceBlock.java b/src/main/java/net/minecraft/world/level/block/IceBlock.java
index 04089e6f7d6e3d532b00585870283922b6be5246..7b7dd00d183aea109f5aaf0bae641734e50c083a 100644
--- a/src/main/java/net/minecraft/world/level/block/IceBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/IceBlock.java
@@ -49,6 +49,10 @@ public class IceBlock extends HalfTransparentBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (world.getBrightness(LightLayer.BLOCK, pos) > 11 - state.getLightBlock(world, pos)) {
             this.melt(state, world, pos);
         }
diff --git a/src/main/java/net/minecraft/world/level/block/LeavesBlock.java b/src/main/java/net/minecraft/world/level/block/LeavesBlock.java
index ebee1dfdb1d75381de847834262a66b45fd9f0ce..e377fed577f8ce96864c4f889c00998679cec4e8 100644
--- a/src/main/java/net/minecraft/world/level/block/LeavesBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LeavesBlock.java
@@ -45,11 +45,19 @@ public class LeavesBlock extends Block implements SimpleWaterloggedBlock {
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return (Integer) state.getValue(LeavesBlock.DISTANCE) == 7 && !(Boolean) state.getValue(LeavesBlock.PERSISTENT);
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (this.decaying(state)) {
             // CraftBukkit start
             LeavesDecayEvent event = new LeavesDecayEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
diff --git a/src/main/java/net/minecraft/world/level/block/LiquidBlock.java b/src/main/java/net/minecraft/world/level/block/LiquidBlock.java
index 4a1830f85f47014da63e4584f411d13f0f0cd8b3..b779235fad70b40e27779905be9f2bcb03ebf279 100644
--- a/src/main/java/net/minecraft/world/level/block/LiquidBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LiquidBlock.java
@@ -60,11 +60,19 @@ public class LiquidBlock extends Block implements BucketPickup {
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return state.getFluidState().isRandomlyTicking();
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         state.getFluidState().randomTick(world, pos, random);
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/MangrovePropaguleBlock.java b/src/main/java/net/minecraft/world/level/block/MangrovePropaguleBlock.java
index cf240750aecc57d287efc986b471e66a860850ee..43cc46284efad11e5056f35028fff84bc99d984f 100644
--- a/src/main/java/net/minecraft/world/level/block/MangrovePropaguleBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/MangrovePropaguleBlock.java
@@ -88,6 +88,10 @@ public class MangrovePropaguleBlock extends SaplingBlock implements SimpleWaterl
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (!isHanging(state)) {
             if (random.nextFloat() < (world.spigotConfig.saplingModifier / (100.0F * 7))) { // Paper
                 this.advanceTree(world, pos, state, random);
diff --git a/src/main/java/net/minecraft/world/level/block/MushroomBlock.java b/src/main/java/net/minecraft/world/level/block/MushroomBlock.java
index 302c5a6401facf192677b89cc0e9190bb35b1229..e9f29e0c1b6c64262cbd50954ff5e181228a6b3f 100644
--- a/src/main/java/net/minecraft/world/level/block/MushroomBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/MushroomBlock.java
@@ -39,6 +39,10 @@ public class MushroomBlock extends BushBlock implements BonemealableBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (random.nextFloat() < (world.spigotConfig.mushroomModifier / (100.0f * 25))) { // Spigot - SPIGOT-7159: Better modifier resolution
             int i = 5;
             boolean flag = true;
diff --git a/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java b/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
index a6ab0d0defc05e56a91084c49897059670a1324b..b7416108d8c5890db5aa3a32a9b629a47aac1424 100644
--- a/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -52,6 +52,10 @@ public class NetherPortalBlock extends Block {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (world.spigotConfig.enableZombiePigmenPortalSpawns && world.dimensionType().natural() && world.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && random.nextInt(2000) < world.getDifficulty().getId()) { // Spigot
             while (world.getBlockState(pos).is((Block) this)) {
                 pos = pos.below();
diff --git a/src/main/java/net/minecraft/world/level/block/NetherWartBlock.java b/src/main/java/net/minecraft/world/level/block/NetherWartBlock.java
index e55720c4d2fbdf6aae526910e87a67c29cf906fd..1d850147fa4d023a40a0bce4ee38e93b5879ccd6 100644
--- a/src/main/java/net/minecraft/world/level/block/NetherWartBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NetherWartBlock.java
@@ -37,11 +37,19 @@ public class NetherWartBlock extends BushBlock {
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return (Integer) state.getValue(NetherWartBlock.AGE) < 3;
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         int i = (Integer) state.getValue(NetherWartBlock.AGE);
 
         if (i < 3 && random.nextFloat() < (world.spigotConfig.wartModifier / (100.0f * 10))) { // Spigot - SPIGOT-7159: Better modifier resolution
diff --git a/src/main/java/net/minecraft/world/level/block/NyliumBlock.java b/src/main/java/net/minecraft/world/level/block/NyliumBlock.java
index 47a3a9e0f0ba4f8a00828874e73d501d18f0ae1c..a5dcc1970c66bd7d8e1eaef4112c1b97e42c359a 100644
--- a/src/main/java/net/minecraft/world/level/block/NyliumBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NyliumBlock.java
@@ -32,6 +32,10 @@ public class NyliumBlock extends Block implements BonemealableBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (!NyliumBlock.canBeNylium(state, world, pos)) {
             // CraftBukkit start
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(world, pos, Blocks.NETHERRACK.defaultBlockState()).isCancelled()) {
diff --git a/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java b/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java
index 1641453fd17938bc8fcdba7f89a0515ebeef20c5..89f91ef083103f0b6987a4631b7976a5d4a77963 100644
--- a/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/PitcherCropBlock.java
@@ -106,11 +106,19 @@ public class PitcherCropBlock extends DoublePlantBlock implements BonemealableBl
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return state.getValue(HALF) == DoubleBlockHalf.LOWER && !this.isMaxAge(state);
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         float f = CropBlock.getGrowthSpeed(this, world, pos);
         boolean bl = random.nextFloat() < (world.spigotConfig.pitcherPlantModifier / (100.0F * (Math.floor(25.0F / f) + 1))); // Paper
         if (bl) {
diff --git a/src/main/java/net/minecraft/world/level/block/PointedDripstoneBlock.java b/src/main/java/net/minecraft/world/level/block/PointedDripstoneBlock.java
index cd943997f11f5ea5c600fdc6db96043fb0fa713c..2cb82df7bafde64495603326b633b4a40347b3e4 100644
--- a/src/main/java/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -177,6 +177,10 @@ public class PointedDripstoneBlock extends Block implements Fallable, SimpleWate
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         PointedDripstoneBlock.maybeTransferFluid(state, world, pos, random.nextFloat());
         if (random.nextFloat() < 0.011377778F && PointedDripstoneBlock.isStalactiteStartPos(state, world, pos)) {
             PointedDripstoneBlock.growStalactiteOrStalagmiteIfPossible(state, world, pos, random);
diff --git a/src/main/java/net/minecraft/world/level/block/RedStoneOreBlock.java b/src/main/java/net/minecraft/world/level/block/RedStoneOreBlock.java
index 06760ff3d20bcfdc30f3f3ee65bc2a0d12d28aff..92816e777ddc20f0020e3bc4dfb33702405ba75d 100644
--- a/src/main/java/net/minecraft/world/level/block/RedStoneOreBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -90,11 +90,19 @@ public class RedStoneOreBlock extends Block {
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return (Boolean) state.getValue(RedStoneOreBlock.LIT);
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if ((Boolean) state.getValue(RedStoneOreBlock.LIT)) {
             // CraftBukkit start
             if (CraftEventFactory.callBlockFadeEvent(world, pos, state.setValue(RedStoneOreBlock.LIT, false)).isCancelled()) {
diff --git a/src/main/java/net/minecraft/world/level/block/SaplingBlock.java b/src/main/java/net/minecraft/world/level/block/SaplingBlock.java
index 53ac4e618fec3fe384d8a106c521f3eace0b5b35..f8f5301bd8327fb79e0a090755966b3c67a212a8 100644
--- a/src/main/java/net/minecraft/world/level/block/SaplingBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SaplingBlock.java
@@ -42,10 +42,13 @@ public class SaplingBlock extends BushBlock implements BonemealableBlock {
 
     @Override
     public void randomTick(net.minecraft.world.level.block.state.BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (world.getMaxLocalRawBrightness(pos.above()) >= 9 && random.nextFloat() < (world.spigotConfig.saplingModifier / (100.0f * 7))) { // Spigot - SPIGOT-7159: Better modifier resolution
             this.advanceTree(world, pos, state, random);
         }
-
     }
 
     public void advanceTree(ServerLevel world, BlockPos pos, net.minecraft.world.level.block.state.BlockState state, RandomSource random) {
diff --git a/src/main/java/net/minecraft/world/level/block/SnowLayerBlock.java b/src/main/java/net/minecraft/world/level/block/SnowLayerBlock.java
index 14e00c7feb1c051d56a3d27cd00dcef072dd771a..fbe73840c1467712a8010122bfd64885ead9fe5b 100644
--- a/src/main/java/net/minecraft/world/level/block/SnowLayerBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SnowLayerBlock.java
@@ -91,6 +91,10 @@ public class SnowLayerBlock extends Block {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (world.getBrightness(LightLayer.BLOCK, pos) > 11) {
             // CraftBukkit start
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(world, pos, Blocks.AIR.defaultBlockState()).isCancelled()) {
diff --git a/src/main/java/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java b/src/main/java/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
index 9bbb9f8e917288bb0d11661a1399a05631ebcce0..c408d0101eda47fd3f59b4de6a1016e359a68364 100644
--- a/src/main/java/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
@@ -51,6 +51,10 @@ public abstract class SpreadingSnowyDirtBlock extends SnowyDirtBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (this instanceof GrassBlock && world.paperConfig().tickRates.grassSpread != 1 && (world.paperConfig().tickRates.grassSpread < 1 || (MinecraftServer.currentTick + pos.hashCode()) % world.paperConfig().tickRates.grassSpread != 0)) { return; } // Paper
         // Paper start
         net.minecraft.world.level.chunk.ChunkAccess cachedBlockChunk = world.getChunkIfLoaded(pos);
diff --git a/src/main/java/net/minecraft/world/level/block/StairBlock.java b/src/main/java/net/minecraft/world/level/block/StairBlock.java
index 827713a332baa6d280cd61d40faf8f6243b3b08c..1e1d97bdab149aa42e4c11cf69d26b35eefef97a 100644
--- a/src/main/java/net/minecraft/world/level/block/StairBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/StairBlock.java
@@ -144,11 +144,19 @@ public class StairBlock extends Block implements SimpleWaterloggedBlock {
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return this.base.isRandomlyTicking(state);
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         this.base.randomTick(state, world, pos, random);
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/StemBlock.java b/src/main/java/net/minecraft/world/level/block/StemBlock.java
index 0d9be3b28ba4957c3f4da5455283fca903836c91..5bc6ac35589f7d82252a03ba42ce9bc4d785abf4 100644
--- a/src/main/java/net/minecraft/world/level/block/StemBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/StemBlock.java
@@ -50,6 +50,10 @@ public class StemBlock extends BushBlock implements BonemealableBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (world.getRawBrightness(pos, 0) >= 9) {
             float f = CropBlock.getGrowthSpeed(this, world, pos);
 
diff --git a/src/main/java/net/minecraft/world/level/block/SugarCaneBlock.java b/src/main/java/net/minecraft/world/level/block/SugarCaneBlock.java
index c3f500580d257e1397f2eb7c47b063a6fe6bb405..f7d7e8c11e23ae2605adf029df0b692c4aac545f 100644
--- a/src/main/java/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -45,6 +45,10 @@ public class SugarCaneBlock extends Block {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (world.isEmptyBlock(pos.above())) {
             int i;
 
diff --git a/src/main/java/net/minecraft/world/level/block/SweetBerryBushBlock.java b/src/main/java/net/minecraft/world/level/block/SweetBerryBushBlock.java
index 34eb7ba1adb51e394bf46a6f643db3529626d9ec..0f391076db8bbfe7ad5c8fba30167c6492d3c53e 100644
--- a/src/main/java/net/minecraft/world/level/block/SweetBerryBushBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SweetBerryBushBlock.java
@@ -59,11 +59,19 @@ public class SweetBerryBushBlock extends BushBlock implements BonemealableBlock
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return (Integer) state.getValue(SweetBerryBushBlock.AGE) < 3;
     }
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         int i = (Integer) state.getValue(SweetBerryBushBlock.AGE);
 
         if (i < 3 && random.nextFloat() < (world.spigotConfig.sweetBerryModifier / (100.0f * 5)) && world.getRawBrightness(pos.above(), 0) >= 9) { // Spigot - SPIGOT-7159: Better modifier resolution
diff --git a/src/main/java/net/minecraft/world/level/block/TorchflowerCropBlock.java b/src/main/java/net/minecraft/world/level/block/TorchflowerCropBlock.java
index 092759184adbf27ce942a148169e60d3a28c2cc7..2678e71bc1f3d28d199956c69668a2fe180daccf 100644
--- a/src/main/java/net/minecraft/world/level/block/TorchflowerCropBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TorchflowerCropBlock.java
@@ -58,6 +58,10 @@ public class TorchflowerCropBlock extends CropBlock {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (random.nextInt(3) != 0) {
             super.randomTick(state, world, pos, random);
         }
diff --git a/src/main/java/net/minecraft/world/level/block/TurtleEggBlock.java b/src/main/java/net/minecraft/world/level/block/TurtleEggBlock.java
index 07157822659288e76b2fba34db8d5d566506f33b..5ec340f28936b836b3310e22fd6065dfbf7477ed 100644
--- a/src/main/java/net/minecraft/world/level/block/TurtleEggBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TurtleEggBlock.java
@@ -105,6 +105,10 @@ public class TurtleEggBlock extends Block {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (this.shouldUpdateHatchLevel(world) && TurtleEggBlock.onSand(world, pos)) {
             int i = (Integer) state.getValue(TurtleEggBlock.HATCH);
 
diff --git a/src/main/java/net/minecraft/world/level/block/VineBlock.java b/src/main/java/net/minecraft/world/level/block/VineBlock.java
index 698ef6ee1c8d08bb38e2775804c1333222b0d73d..b139803946cb8e3f0f6ea983edbefb8412d6b982 100644
--- a/src/main/java/net/minecraft/world/level/block/VineBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/VineBlock.java
@@ -177,6 +177,10 @@ public class VineBlock extends Block {
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         if (world.getGameRules().getBoolean(GameRules.RULE_DO_VINES_SPREAD)) {
             if (random.nextFloat() < (world.spigotConfig.vineModifier / (100.0f * 4))) { // Spigot - SPIGOT-7159: Better modifier resolution
                 Direction enumdirection = Direction.getRandom(random);
diff --git a/src/main/java/net/minecraft/world/level/block/WeatheringCopperFullBlock.java b/src/main/java/net/minecraft/world/level/block/WeatheringCopperFullBlock.java
index 05cccc33e7745eb3a905a804b318aaa6bd440d04..2278bff4545cb258cb076422a4c7589527c39ad2 100644
--- a/src/main/java/net/minecraft/world/level/block/WeatheringCopperFullBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/WeatheringCopperFullBlock.java
@@ -16,11 +16,19 @@ public class WeatheringCopperFullBlock extends Block implements WeatheringCopper
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         this.onRandomTick(state, world, pos, random);
     }
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return WeatheringCopper.getNext(state.getBlock()).isPresent();
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/WeatheringCopperSlabBlock.java b/src/main/java/net/minecraft/world/level/block/WeatheringCopperSlabBlock.java
index 37d7fb85a05521e4acb6a79997b661c1895bb7ed..7f3ca01da48db936534320c85f1efcb2b8bea96c 100644
--- a/src/main/java/net/minecraft/world/level/block/WeatheringCopperSlabBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/WeatheringCopperSlabBlock.java
@@ -16,11 +16,19 @@ public class WeatheringCopperSlabBlock extends SlabBlock implements WeatheringCo
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         this.onRandomTick(state, world, pos, random);
     }
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return WeatheringCopper.getNext(state.getBlock()).isPresent();
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/WeatheringCopperStairBlock.java b/src/main/java/net/minecraft/world/level/block/WeatheringCopperStairBlock.java
index 728a607bb28f10955397744970554efe297ea5ca..ef7ce2d63017d133bef53aa1b721be46562ac53e 100644
--- a/src/main/java/net/minecraft/world/level/block/WeatheringCopperStairBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/WeatheringCopperStairBlock.java
@@ -16,11 +16,19 @@ public class WeatheringCopperStairBlock extends StairBlock implements Weathering
 
     @Override
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         this.onRandomTick(state, world, pos, random);
     }
 
     @Override
     public boolean isRandomlyTicking(BlockState state) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomTickBehaviourImplemented(state))
+            return de.verdox.mccreativelab.behaviour.BehaviourUtil.isRandomlyTicking(state);
+        // MCCLab - end
         return WeatheringCopper.getNext(state.getBlock()).isPresent();
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
index e6a4a5898ffdcb2aa2bc01371a6d7dbc06d610ce..3c8e09b76407d0782109b172765a97454ba9a7d3 100644
--- a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -349,6 +349,10 @@ public abstract class BlockBehaviour implements FeatureElement {
     /** @deprecated */
     @Deprecated
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+        // MCCLab - start
+        if(de.verdox.mccreativelab.behaviour.BehaviourUtil.randomTick(state, world, pos, random))
+            return;
+        // MCCLab - end
         this.tick(state, world, pos, random);
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b4077b0f4da2ee2e2aaffb1331a86e0580bdee3a..3f11183fe7d6ceb365895e83b18ec4dcc5863518 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1096,6 +1096,7 @@ public final class CraftServer implements Server {
         io.papermc.paper.plugin.PluginInitializerManager.reload(this.console); // Paper
         this.loadPlugins();
         this.enablePlugins(PluginLoadOrder.STARTUP);
+        net.minecraft.world.level.block.Blocks.rebuildCache();
         this.enablePlugins(PluginLoadOrder.POSTWORLD);
         this.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.RELOAD));
         org.spigotmc.WatchdogThread.hasStarted = true; // Paper - Disable watchdog early timeout on reload
