From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Fri, 31 Mar 2023 16:20:01 +0200
Subject: [PATCH] Allowing custom registries by skipping unknown registries on
 reload


diff --git a/src/main/java/net/minecraft/server/ReloadableServerResources.java b/src/main/java/net/minecraft/server/ReloadableServerResources.java
index 3192ad97889f1a199b4b7b3dc43ff9bbafe748b3..b38bd8ab26868b122e982c40e231d3987dc5afb6 100644
--- a/src/main/java/net/minecraft/server/ReloadableServerResources.java
+++ b/src/main/java/net/minecraft/server/ReloadableServerResources.java
@@ -90,6 +90,9 @@ public class ReloadableServerResources {
         }, (entry) -> {
             return List.copyOf(entry.getValue());
         }));
-        dynamicRegistryManager.registryOrThrow(resourceKey).bindTags(map);
+
+        dynamicRegistryManager.registry(resourceKey).ifPresent(ts -> ts.bindTags(map));
+
+        //dynamicRegistryManager.registryOrThrow(resourceKey).bindTags(map);
     }
 }
