From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Wed, 1 May 2024 15:15:21 +0000
Subject: [PATCH] Adding ItemDisplayBehaviour


diff --git a/src/main/java/de/verdox/mccreativelab/behaviour/ItemDisplayBehaviourUtil.java b/src/main/java/de/verdox/mccreativelab/behaviour/ItemDisplayBehaviourUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..8638caba3f51dadd2f3e92778ad0fc76285d2b77
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/behaviour/ItemDisplayBehaviourUtil.java
@@ -0,0 +1,25 @@
+package de.verdox.mccreativelab.behaviour;
+
+import io.papermc.paper.adventure.PaperAdventure;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.Style;
+import net.minecraft.world.item.ItemStack;
+
+public class ItemDisplayBehaviourUtil extends BehaviourUtil {
+    public static Style basicLoreStyle(ItemStack nmsStack){
+        if(!ItemDisplayBehaviour.BASIC_LORE_BEHAVIOUR.isImplemented())
+            return Style.EMPTY.withColor(ChatFormatting.DARK_PURPLE).withItalic(true);
+
+        net.kyori.adventure.text.format.Style style = ItemDisplayBehaviour.BASIC_LORE_BEHAVIOUR.getBehaviour().basicLoreStyle(nmsStack.getBukkitStack());
+        return PaperAdventure.asVanilla(style);
+    }
+
+    public static Style basicNameStyle(ItemStack nmsStack){
+        if(!ItemDisplayBehaviour.BASIC_LORE_BEHAVIOUR.isImplemented())
+            return Style.EMPTY.withItalic(true);
+
+        net.kyori.adventure.text.format.Style style = ItemDisplayBehaviour.BASIC_LORE_BEHAVIOUR.getBehaviour().basicCustomNameStyle(nmsStack.getBukkitStack());
+        return PaperAdventure.asVanilla(style);
+    }
+
+}
diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 512f4d80ebab6fe153767c9265cf5d2e07dcfe56..63c06489923e1ebf0d6ea907016c7ef441a0c0a2 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -994,7 +994,7 @@ public final class ItemStack {
         MutableComponent ichatmutablecomponent = Component.empty().append(this.getHoverName()).withStyle(this.getRarity().color);
 
         if (this.hasCustomHoverName()) {
-            ichatmutablecomponent.withStyle(ChatFormatting.ITALIC);
+            ichatmutablecomponent.withStyle(de.verdox.mccreativelab.behaviour.ItemDisplayBehaviourUtil.basicNameStyle(this));
         }
 
         list.add(ichatmutablecomponent);
@@ -1044,7 +1044,7 @@ public final class ItemStack {
                             MutableComponent ichatmutablecomponent1 = Component.Serializer.fromJson(s);
 
                             if (ichatmutablecomponent1 != null) {
-                                list.add(ComponentUtils.mergeStyles(ichatmutablecomponent1, ItemStack.LORE_STYLE));
+                                list.add(ComponentUtils.mergeStyles(ichatmutablecomponent1, de.verdox.mccreativelab.behaviour.ItemDisplayBehaviourUtil.basicLoreStyle(this)));
                             }
                         } catch (Exception exception) {
                             nbttagcompound.remove("Lore");
@@ -1322,7 +1322,7 @@ public final class ItemStack {
         MutableComponent ichatmutablecomponent = Component.empty().append(this.getHoverName());
 
         if (this.hasCustomHoverName()) {
-            ichatmutablecomponent.withStyle(ChatFormatting.ITALIC);
+            ichatmutablecomponent.withStyle(de.verdox.mccreativelab.behaviour.ItemDisplayBehaviourUtil.basicNameStyle(this));
         }
 
         MutableComponent ichatmutablecomponent1 = ComponentUtils.wrapInSquareBrackets(ichatmutablecomponent);
