From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Wed, 15 Nov 2023 17:47:39 +0100
Subject: [PATCH] Adding growTreeIfSapling helper method to Bukkit World class


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 33e4818ba5a90d78d69baad9f6b1be1b1382e9f3..95c7838832f988ab32bbd4c92482911024edb59e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -199,6 +199,21 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     }
     // Paper end - void damage configuration
 
+    // MCCLab - start
+    @Override
+    public boolean growTreeIfSapling(@NotNull Location location) {
+        var blockPos = new BlockPos(location.getBlockX(),location.getBlockY(),location.getBlockZ());
+        var block = world.getBlockIfLoaded(blockPos);
+        if(block == null)
+            return false;
+        var blockState = world.getBlockState(blockPos);
+        if(block instanceof net.minecraft.world.level.block.SaplingBlock saplingBlock) {
+            saplingBlock.advanceTree(world, new BlockPos(location.getBlockX(), location.getBlockY(), location.getBlockZ()), blockState, world.random);
+            return true;
+        }
+        return false;
+    }
+    // MCCLab - end
     // Paper start - Provide fast information methods
     @Override
     public int getEntityCount() {
