From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Wed, 1 May 2024 15:09:35 +0000
Subject: [PATCH] Add brokenState to BlockBreakEvent


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index 352a5ccae6ae25d84033f968b1385a86e2e9c4c7..481c16b1deffb926ae23ada2a2208548c9f64b1a 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -357,7 +357,7 @@ public class ServerPlayerGameMode {
                 this.player.connection.send(packet);
             }
 
-            event = new BlockBreakEvent(bblock, this.player.getBukkitEntity());
+            event = new BlockBreakEvent(bblock, bblock.getState(), this.player.getBukkitEntity());
 
             // Sword + Creative mode pre-cancel
             event.setCancelled(isSwordNoBreak);
