From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sun, 3 Dec 2023 14:20:30 +0100
Subject: [PATCH] Implementing Basic Item Format


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 5b8a1f31e0b55da15daa4ab271317e4393a87e96..3f0b0ef9e4b788828785a718ea6c2adc43ab8cfd 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -246,6 +246,7 @@ public final class ItemStack {
     public ItemStack(Holder<Item> item, int count, Optional<CompoundTag> nbt) {
         this(item, count);
         nbt.ifPresent(this::setTag);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     public ItemStack(Holder<Item> itemEntry, int count) {
@@ -258,7 +259,7 @@ public final class ItemStack {
         if (this.item.canBeDepleted()) {
             this.setDamageValue(this.getDamageValue());
         }
-
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     private ItemStack(@Nullable Void v) {
@@ -294,6 +295,7 @@ public final class ItemStack {
 
     private ItemStack(CompoundTag nbt) {
         this.load(nbt);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
         // CraftBukkit end
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index c1d902c8dd2ec23240ee60ca9e9eaf7c839baed5..0a4c2aae632dbef1998b1e77cfd18f793989fd33 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -129,6 +129,7 @@ public final class CraftItemStack extends ItemStack {
         this.setAmount(amount);
         this.setDurability(durability);
         this.setItemMeta(itemMeta);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(this);
     }
 
     @Override
