From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Fri, 31 Mar 2023 16:20:01 +0200
Subject: [PATCH] Allowing custom registries by skipping unknown registries on
 reload


diff --git a/src/main/java/net/minecraft/server/ReloadableServerResources.java b/src/main/java/net/minecraft/server/ReloadableServerResources.java
index 3192ad97889f1a199b4b7b3dc43ff9bbafe748b3..8f5499c359cc1910da68482d5b95f5a75a67a64d 100644
--- a/src/main/java/net/minecraft/server/ReloadableServerResources.java
+++ b/src/main/java/net/minecraft/server/ReloadableServerResources.java
@@ -90,6 +90,6 @@ public class ReloadableServerResources {
         }, (entry) -> {
             return List.copyOf(entry.getValue());
         }));
-        dynamicRegistryManager.registryOrThrow(resourceKey).bindTags(map);
+        dynamicRegistryManager.registry(resourceKey).ifPresent(ts -> ts.bindTags(map)); // MCCLab - Dont throw exception on unknown registries
     }
 }
