From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Thu, 23 Nov 2023 00:41:02 +0100
Subject: [PATCH] Adding Utility class for MCCreativeLab


diff --git a/src/main/java/de/verdox/mccreativelab/CreativeLabImpl.java b/src/main/java/de/verdox/mccreativelab/CreativeLabImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..863d73d799e9838534faea0254d5b37e7f7748cb
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/CreativeLabImpl.java
@@ -0,0 +1,5 @@
+package de.verdox.mccreativelab;
+
+public final class CreativeLabImpl implements CreativeLab {
+
+}
diff --git a/src/main/java/net/minecraft/server/Bootstrap.java b/src/main/java/net/minecraft/server/Bootstrap.java
index 8c2f9f549b5155b8d8fecbc22164b334805381e2..642fe67528b647fd6881820f69d7760ec88d4091 100644
--- a/src/main/java/net/minecraft/server/Bootstrap.java
+++ b/src/main/java/net/minecraft/server/Bootstrap.java
@@ -62,6 +62,11 @@ public class Bootstrap {
             Bootstrap.isBootstrapped = true;
             Instant instant = Instant.now();
 
+            // MCCLab - start
+            de.verdox.mccreativelab.MCCreativeLab.setCreativeLab(new de.verdox.mccreativelab.CreativeLabImpl());
+            MinecraftServer.LOGGER.info("MCCreativeLab: initialized");
+            // MCCLab - end
+
             io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enterBootstrappers(); // Paper - Entrypoint for bootstrapping
             if (BuiltInRegistries.REGISTRY.keySet().isEmpty()) {
                 throw new IllegalStateException("Unable to load registries");
