From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Sun, 30 Jun 2024 18:49:04 +0000
Subject: [PATCH] Fixing BukkitVersion


diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index 7f4d220c3a689e46ccaad5a1f79c1fc24057e59e..705a604ac226a385b28304e0833614050d1bb016 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -11,7 +11,7 @@ public final class Versioning {
     public static String getBukkitVersion() {
         String result = "Unknown-Version";
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/com.example.paperfork/mccreativelab-api/pom.properties"); // ForkTest
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/de.verdox.mccreativelab/mccreativelab-api/pom.properties"); // ForkTest
         Properties properties = new Properties();
 
         if (stream != null) {
