From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Thu, 8 Aug 2024 11:58:55 +0000
Subject: [PATCH] Compiler error fix


diff --git a/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
index c90368dff20befa244a737aaf4ee49b11c710258..f7574538c38331fc0616e1e6f30e4e216feff162 100644
--- a/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
+++ b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
@@ -20,8 +20,6 @@ public class CraftPersistentDataContainer extends io.papermc.paper.persistence.P
 
     private final Map<String, Tag> customDataTags = new HashMap<>();
     // Paper - move to PersistentDataContainerView
-    private final CraftPersistentDataTypeRegistry registry;
-    private final CraftPersistentDataAdapterContext adapterContext;
     private final CraftPersistentDataObjectCache craftPersistentDataObjectCache; // Paper - Adding PersistentDataObjects
 
     public CraftPersistentDataContainer(Map<String, Tag> customTags, CraftPersistentDataTypeRegistry registry) {
