--- a/fork-server/build.gradle.kts
+++ b/fork-server/build.gradle.kts
@@ -33,8 +_,22 @@
         }
     }
 
-    activeFork = fork
+    val forky = forks.register("forky") {
+        forks = fork
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = fork.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("forky-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+        upstream.patchDir("forkServer") {
+            upstreamPath = "fork-server"
+            excludes = setOf("src/minecraft", "paper-patches", "minecraft-patches", "build.gradle.kts", "build.gradle.kts.patch")
+            patchesDir = rootDirectory.dir("forky-server/fork-patches")
+            outputDir = rootDirectory.dir("fork-server")
+        }
+    }
 
+    activeFork = forky
 
     //updatingMinecraft {
     //    oldPaperCommit = "f4f275519f7c1fbe9db173b7144a4fe81440e365"
@@ -124,10 +_,14 @@
     main {
         java { srcDir("../paper-server/src/main/java") }
         resources { srcDir("../paper-server/src/main/resources") }
+        java { srcDir("../fork-server/src/main/java") }
+        resources { srcDir("../fork-server/src/main/resources") }
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
+        java { srcDir("../fork-server/src/test/java") }
+        resources { srcDir("../fork-server/src/test/resources") }
     }
 }
 
@@ -155,7 +_,7 @@
 }
 
 dependencies {
-    implementation(project(":fork-api"))
+    implementation(project(":forky-api"))
     implementation("ca.spottedleaf:concurrentutil:0.0.3")
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
