From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sun, 17 Mar 2024 00:35:46 +0100
Subject: [PATCH] Adding getAllBlockDataVariants metod


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 335dad2d2687982354831ccb9fe7208d28699187..423615c127fb46b37ad0d986b43372fbe84d54df 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2429,6 +2429,17 @@ public final class Bukkit {
         return server.advancementIterator();
     }
 
+    // Paper start - Adding getAllBlockDataVariants methods
+    /**
+     * Returns all block data variants of a material block
+     * @param material - The Block material
+     * @return - The List of block data variants
+     */
+    public static List<BlockData> getAllBlockDataVariants(Material material){
+        return server.getAllBlockDataVariants(material);
+    }
+    // Paper start - Adding getAllBlockDataVariants methods
+
     /**
      * Creates a new {@link BlockData} instance for the specified Material, with
      * all properties initialized to unspecified defaults.
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9bd65d894412d9705e5effdb0b81129b29551ae3..2fa7ab7931bd6dd4e1ab5613e7f2d684502dd877 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2087,6 +2087,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     Iterator<Advancement> advancementIterator();
 
+    // Paper start - Adding getAllBlockDataVariants methods
+    /**
+     * Returns all block data variants of a material block
+     * @param material - The Block material
+     * @return - The List of block data variants
+     */
+    List<BlockData> getAllBlockDataVariants(Material material);
+    // Paper start - Adding getAllBlockDataVariants methods
+
     /**
      * Creates a new {@link BlockData} instance for the specified Material, with
      * all properties initialized to unspecified defaults.
