From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Thu, 20 Jun 2024 13:25:07 +0000
Subject: [PATCH] Fix Insert missing imports


diff --git a/src/main/java/de/verdox/mccreativelab/events/ChunkDataCreateEvent.java b/src/main/java/de/verdox/mccreativelab/events/ChunkDataCreateEvent.java
index 8df6c57721400f106daf43e20768e585c3f4e19b..25f7616f57b33d8efb8c93f86466548b0ebf4d67 100644
--- a/src/main/java/de/verdox/mccreativelab/events/ChunkDataCreateEvent.java
+++ b/src/main/java/de/verdox/mccreativelab/events/ChunkDataCreateEvent.java
@@ -1,6 +1,7 @@
 package de.verdox.mccreativelab.events;
 
 import org.bukkit.World;
+import org.bukkit.event.HandlerList;
 import org.bukkit.persistence.PersistentDataContainer;
 import org.jetbrains.annotations.NotNull;
 import de.verdox.mccreativelab.worldgen.WorldGenChunk;
diff --git a/src/main/java/de/verdox/mccreativelab/events/ChunkDataLoadEvent.java b/src/main/java/de/verdox/mccreativelab/events/ChunkDataLoadEvent.java
index 64011b084ccafbe53970a34fb3432776f2a1f11c..0150cc269901331d4895282ea1f2253b198533bf 100644
--- a/src/main/java/de/verdox/mccreativelab/events/ChunkDataLoadEvent.java
+++ b/src/main/java/de/verdox/mccreativelab/events/ChunkDataLoadEvent.java
@@ -2,6 +2,7 @@ package de.verdox.mccreativelab.events;
 
 import org.bukkit.Chunk;
 import org.bukkit.World;
+import org.bukkit.event.HandlerList;
 import org.bukkit.persistence.PersistentDataContainer;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
diff --git a/src/main/java/de/verdox/mccreativelab/events/ChunkDataSaveEvent.java b/src/main/java/de/verdox/mccreativelab/events/ChunkDataSaveEvent.java
index 8418bc26de6b3440df443b55c5b8b9b942f4f33c..563d0da12c606ee80ac7d88edbb23d02214890fb 100644
--- a/src/main/java/de/verdox/mccreativelab/events/ChunkDataSaveEvent.java
+++ b/src/main/java/de/verdox/mccreativelab/events/ChunkDataSaveEvent.java
@@ -2,6 +2,7 @@ package de.verdox.mccreativelab.events;
 
 import org.bukkit.Chunk;
 import org.bukkit.World;
+import org.bukkit.event.HandlerList;
 import org.bukkit.persistence.PersistentDataContainer;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
