From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Fri, 17 Nov 2023 19:02:10 +0100
Subject: [PATCH] Opening fuel burn durations to API


diff --git a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index a18aadbf7ae83713e1f2b21553185d8000bc7699..a3c46f29d95d71913406217cda8666cdb4bca90e 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -478,6 +478,8 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
     }
 
     protected int getBurnDuration(ItemStack fuel) {
+        if(de.verdox.mccreativelab.behaviour.FuelBurnBehaviour.FUEL_BURN_BEHAVIOUR.isImplemented())
+            return de.verdox.mccreativelab.behaviour.FuelBurnBehaviour.FUEL_BURN_BEHAVIOUR.getBehaviour().calculateFurnaceBurnDuration(CraftItemStack.asCraftMirror(fuel), getLocation().getBlock());
         if (fuel.isEmpty()) {
             return 0;
         } else {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/BlastFurnaceBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BlastFurnaceBlockEntity.java
index fb4f57b2ed17ff9b4853e1ba2aac7a002989e5c8..e41cb23da764a3008de99e40e8dbe29f36a7d71b 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BlastFurnaceBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BlastFurnaceBlockEntity.java
@@ -21,6 +21,8 @@ public class BlastFurnaceBlockEntity extends AbstractFurnaceBlockEntity {
 
     @Override
     protected int getBurnDuration(ItemStack fuel) {
+        if(de.verdox.mccreativelab.behaviour.FuelBurnBehaviour.FUEL_BURN_BEHAVIOUR.isImplemented())
+            return de.verdox.mccreativelab.behaviour.FuelBurnBehaviour.FUEL_BURN_BEHAVIOUR.getBehaviour().calculateBlastFurnaceBurnDuration(fuel.getBukkitStack(), getLocation().getBlock());
         return super.getBurnDuration(fuel) / 2;
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/entity/SmokerBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SmokerBlockEntity.java
index 7d9548f89fda9f4ba203d95d6e64826e9a697013..0c4fb2036cdf78576797d0d3d4068ebe1643855e 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SmokerBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SmokerBlockEntity.java
@@ -21,6 +21,8 @@ public class SmokerBlockEntity extends AbstractFurnaceBlockEntity {
 
     @Override
     protected int getBurnDuration(ItemStack fuel) {
+        if(de.verdox.mccreativelab.behaviour.FuelBurnBehaviour.FUEL_BURN_BEHAVIOUR.isImplemented())
+            return de.verdox.mccreativelab.behaviour.FuelBurnBehaviour.FUEL_BURN_BEHAVIOUR.getBehaviour().calculateSmokerBurnDuration(fuel.getBukkitStack(), getLocation().getBlock());
         return super.getBurnDuration(fuel) / 2;
     }
 
