From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sat, 20 Jan 2024 15:49:51 +0100
Subject: [PATCH] Add Behaviour upper class


diff --git a/src/main/java/de/verdox/mccreativelab/behaviour/Behaviour.java b/src/main/java/de/verdox/mccreativelab/behaviour/Behaviour.java
new file mode 100644
index 0000000000000000000000000000000000000000..9dc0f3aff2672c513cb6ba3d5f6c42cc40fb33c0
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/behaviour/Behaviour.java
@@ -0,0 +1,31 @@
+package de.verdox.mccreativelab.behaviour;
+
+public interface Behaviour {
+    default BehaviourResult.Void voidResult(BehaviourResult.Void.Type type) {
+        return new BehaviourResult.Void(type);
+    }
+
+    default BehaviourResult.Void voidResult() {
+        return voidResult(BehaviourResult.Void.Type.REPLACE_VANILLA);
+    }
+
+    default BehaviourResult.Callback done() {
+        return new BehaviourResult.Callback();
+    }
+
+    default BehaviourResult.Bool bool(boolean value, BehaviourResult.Bool.Type type) {
+        return new BehaviourResult.Bool(value, type);
+    }
+
+    default BehaviourResult.Bool bool(boolean value) {
+        return bool(value, BehaviourResult.Bool.Type.REPLACE_VANILLA);
+    }
+
+    default <T> BehaviourResult.Object<T> result(T result, BehaviourResult.Object.Type type){
+        return new BehaviourResult.Object<>(result, type);
+    }
+
+    default <T> BehaviourResult.Object<T> result(T result){
+        return result(result, BehaviourResult.Object.Type.REPLACE_VANILLA);
+    }
+}
