From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Sat, 18 Nov 2023 17:15:16 +0100
Subject: [PATCH] Adding MultiCustomBehaviour to API


diff --git a/src/main/java/de/verdox/mccreativelab/MultiCustomBehaviour.java b/src/main/java/de/verdox/mccreativelab/MultiCustomBehaviour.java
new file mode 100644
index 0000000000000000000000000000000000000000..f3f6272c3e6a56a1ead0ea8484a077ac9d7d21cb
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/MultiCustomBehaviour.java
@@ -0,0 +1,51 @@
+package de.verdox.mccreativelab;
+
+import org.bukkit.Bukkit;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class MultiCustomBehaviour<K, T> {
+    private final String key;
+    private final Map<K, T> storedBehaviour = new HashMap<>();
+
+    public MultiCustomBehaviour(String key) {
+        this.key = key;
+    }
+
+    public String getKey() {
+        return key;
+    }
+
+    /**
+     * Used to set a new behaviour.
+     *
+     * @param behaviour The new implementation
+     */
+    public void setBehaviour(@NotNull K key, @NotNull T behaviour) {
+        Bukkit.getLogger().info(key + ": Using implementation " + behaviour.getClass().getSimpleName() + " for " + key);
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(behaviour);
+        storedBehaviour.put(key, behaviour);
+    }
+
+    @Nullable
+    @ApiStatus.Internal
+    public T getBehaviour(@NotNull K key) {
+        Objects.requireNonNull(key);
+        return storedBehaviour.get(key);
+    }
+
+    /**
+     * Checks if a custom implementation exists
+     * @return true if a custom implementation exists
+     */
+    public boolean isImplemented(@NotNull K key){
+        Objects.requireNonNull(key);
+        return storedBehaviour.containsKey(key);
+    }
+}
