From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Thu, 23 Nov 2023 01:00:21 +0100
Subject: [PATCH] Implementing FakeEnum to get rid of former Enum
 implementations while still keeping api calls intact


diff --git a/src/main/java/de/verdox/mccreativelab/data/FakeEnum.java b/src/main/java/de/verdox/mccreativelab/data/FakeEnum.java
new file mode 100644
index 0000000000000000000000000000000000000000..87f9a7403c5165066743e9ad4bb6ebac8b71aa23
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/data/FakeEnum.java
@@ -0,0 +1,22 @@
+package de.verdox.mccreativelab.data;
+
+import org.bukkit.Keyed;
+import org.bukkit.NamespacedKey;
+import org.bukkit.Registry;
+
+import java.util.List;
+import java.util.Locale;
+
+public interface FakeEnum {
+    static <T extends Keyed> T valueOf(String name, Registry<T> storage){
+        return storage.get(NamespacedKey.minecraft(name.replace(".","_").toLowerCase(Locale.ROOT)));
+    }
+
+    static <T extends Keyed> List<T> values(Registry<T> storage){
+        return storage.stream().toList();
+    }
+
+    static String restoreEnumNameSchemeFromKey(NamespacedKey namespacedKey){
+        return namespacedKey.value().replace(".","_").toUpperCase(Locale.ROOT);
+    }
+}
