From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Wed, 24 Jan 2024 01:29:23 +0100
Subject: [PATCH] Make Chunk Metadatable


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index 72cb2ba120f80462378cf704a04893304ea72947..fc69334de4f82694b4e830400c37021c7685fa46 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -19,7 +19,7 @@ import org.jetbrains.annotations.NotNull;
  * then the chunk will only be generated as far as it needs to provide the
  * requested data.
  */
-public interface Chunk extends PersistentDataHolder {
+public interface Chunk extends PersistentDataHolder, org.bukkit.metadata.Metadatable {
 
     /**
      * Gets the X-coordinate of this chunk
