From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Mon, 20 Nov 2023 18:15:19 +0100
Subject: [PATCH] Opening block breaking behaviour to API


diff --git a/src/main/java/de/verdox/mccreativelab/behaviour/BlockBreakBehaviour.java b/src/main/java/de/verdox/mccreativelab/behaviour/BlockBreakBehaviour.java
new file mode 100644
index 0000000000000000000000000000000000000000..8a282230c1dce1c5e952e8a869c9628967415342
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/behaviour/BlockBreakBehaviour.java
@@ -0,0 +1,29 @@
+package de.verdox.mccreativelab.behaviour;
+
+import de.verdox.mccreativelab.MultiCustomBehaviour;
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.block.data.BlockData;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+
+public interface BlockBreakBehaviour {
+    MultiCustomBehaviour<Material, BlockBreakBehaviour> BLOCK_BREAK_BEHAVIOUR = new MultiCustomBehaviour<>("MCCLab - BlockBreakBehaviour");
+
+    /**
+     * This method is called by the minecraft server to ask whether a certain block data requires specific tools to drop from the loot table.
+     * An example for this is stone needing pickaxes to drop cobblestone.
+     * This method is not used to determine which tool is used.
+     * @param blockData The block data
+     * @return true if it requires special tools
+     */
+    boolean requiresCorrectToolsForDrops(BlockData blockData);
+
+    /**
+     * This method is called by the minecraft server to check whether a particular item stack is the correct tool for an item drop.
+     * @param toolAsItemstack The tool used to break the block
+     * @param blockData The block data of the broken block
+     * @return true if the tool is correct.
+     */
+    boolean isCorrectToolMaterialForDrops(ItemStack toolAsItemstack, BlockData blockData);
+}
