From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Fri, 26 Jan 2024 15:15:53 +0100
Subject: [PATCH] Adds a new function to EntityBrain


diff --git a/src/main/java/de/verdox/mccreativelab/ai/EntityBrain.java b/src/main/java/de/verdox/mccreativelab/ai/EntityBrain.java
index 334d84bdb3809d265777d1adc163404ccf7caf76..e0df00a788dced9e5434a00e0028c7300ad77b8c 100644
--- a/src/main/java/de/verdox/mccreativelab/ai/EntityBrain.java
+++ b/src/main/java/de/verdox/mccreativelab/ai/EntityBrain.java
@@ -19,6 +19,8 @@ public interface EntityBrain<E extends LivingEntity> {
 
     void tick(World world, E entity);
 
+    <U> void addMemoryToBrain(MemoryKey<U> memoryKey);
+
     <U> void eraseMemory(MemoryKey<U> memoryKey);
 
     <U> void setMemory(MemoryKey<U> type, @Nullable U value);
