From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Fri, 26 Jan 2024 15:29:48 +0100
Subject: [PATCH] Fix in WorldEffectEvent


diff --git a/src/main/java/io/papermc/paper/event/world/WorldEffectEvent.java b/src/main/java/io/papermc/paper/event/world/WorldEffectEvent.java
index 2e7964834937d16003314cdbd2c424cd77cf1a41..4b01a8a73c3ce8e5d7df756d00892dd451436806 100644
--- a/src/main/java/io/papermc/paper/event/world/WorldEffectEvent.java
+++ b/src/main/java/io/papermc/paper/event/world/WorldEffectEvent.java
@@ -16,10 +16,10 @@ public class WorldEffectEvent extends WorldEvent implements Cancellable {
     private Location soundLocation;
     private Player except;
     private Effect effect;
-    private int data;
+    private Object data;
     private boolean cancelled;
 
-    public WorldEffectEvent(@NotNull World world, @NotNull Location soundLocation, @Nullable Player except, Effect effect, int data, boolean global) {
+    public WorldEffectEvent(@NotNull World world, @NotNull Location soundLocation, @Nullable Player except, Effect effect, Object data, boolean global) {
         super(world);
         this.soundLocation = soundLocation;
         this.except = except;
@@ -60,7 +60,7 @@ public class WorldEffectEvent extends WorldEvent implements Cancellable {
      *
      * @param data - the new data
      */
-    public void setData(int data) {
+    public void setData(Object data) {
         this.data = data;
     }
 
@@ -108,7 +108,7 @@ public class WorldEffectEvent extends WorldEvent implements Cancellable {
      *
      * @return - The data
      */
-    public int getData() {
+    public Object getData() {
         return data;
     }
 
