From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Mon, 1 Jan 2024 18:06:04 +0100
Subject: [PATCH] Adding LightEngineTracker and BlockLightEvent


diff --git a/src/main/java/de/verdox/mccreativelab/events/BlockLightEvent.java b/src/main/java/de/verdox/mccreativelab/events/BlockLightEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..52766dcaf469e5de789abdd98beff7262747b2ed
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/events/BlockLightEvent.java
@@ -0,0 +1,44 @@
+package de.verdox.mccreativelab.events;
+
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Map;
+
+public class BlockLightEvent extends Event {
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+    private final World world;
+    private final Map<Location, int[]> trackedBlockLightChanges;
+
+    public BlockLightEvent(World world, Map<Location, int[]> trackedBlockLightChanges){
+        super(!Bukkit.isPrimaryThread());
+        this.world = world;
+        this.trackedBlockLightChanges = trackedBlockLightChanges;
+    }
+
+    public World getWorld() {
+        return world;
+    }
+
+    /**
+     * The first value of the int array is the block light. The second value is the skylight.
+     * If one value was not updated it is stored as -1.
+     * @return
+     */
+    public Map<Location, int[]> getTrackedBlockLightChanges() {
+        return trackedBlockLightChanges;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+    @NotNull
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+}
