From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Thu, 21 Mar 2024 17:57:27 +0100
Subject: [PATCH] Adding resendChunksForBiomes function


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index deaf45aecab653efdb8381e714a93871972f433e..d11dbfe1a82320ff76d29a073fc159c73e895b7f 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -2355,6 +2355,12 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @Nullable
     public BiomeProvider getBiomeProvider();
 
+    /**
+     * Resends chunks to players
+     * @param chunks - The chunks to resend
+     */
+    public void resendChunksForBiomes(List<Chunk> chunks);
+
     /**
      * Saves world to disk
      */
