From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Thu, 21 Mar 2024 17:57:27 +0100
Subject: [PATCH] Adding resendChunksForBiomes function


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 8e576b98a81c6a638191cdd726998c052abf35de..3795fd2eff436b149df7a8520af6d07cff7b4740 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -2291,6 +2291,12 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @Nullable
     public BiomeProvider getBiomeProvider();
 
+    /**
+     * Resends chunks to players
+     * @param chunks - The chunks to resend
+     */
+    public void resendChunksForBiomes(List<Chunk> chunks);
+
     /**
      * Saves world to disk
      */
