From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sun, 7 Jan 2024 13:39:48 +0100
Subject: [PATCH] Adding InteractionResult for further use with Behaviours


diff --git a/src/main/java/de/verdox/mccreativelab/InteractionResult.java b/src/main/java/de/verdox/mccreativelab/InteractionResult.java
new file mode 100644
index 0000000000000000000000000000000000000000..0daf40262de8269c3dece903bb8f16519da7ed46
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/InteractionResult.java
@@ -0,0 +1,9 @@
+package de.verdox.mccreativelab;
+
+public enum InteractionResult {
+    SUCCESS,
+    CONSUME,
+    CONSUME_PARTIAL,
+    PASS,
+    FAIL;
+}
