From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sun, 17 Mar 2024 00:35:46 +0100
Subject: [PATCH] Adding getAllBlockDataVariants metod


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 0b47aa638a8945a3eecc6aa8bd22ab5e5906b4d5..c1d75a9f83edf174295115033024c2c2a5d5f342 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2421,6 +2421,17 @@ public final class Bukkit {
         return server.advancementIterator();
     }
 
+    // Paper start - Adding getAllBlockDataVariants methods
+    /**
+     * Returns all block data variants of a material block
+     * @param material - The Block material
+     * @return - The List of block data variants
+     */
+    public static List<BlockData> getAllBlockDataVariants(Material material){
+        return server.getAllBlockDataVariants(material);
+    }
+    // Paper start - Adding getAllBlockDataVariants methods
+
     /**
      * Creates a new {@link BlockData} instance for the specified Material, with
      * all properties initialized to unspecified defaults.
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 11741b89bfdf9a1a68ab1e76de84ce7a8119b4e0..b9ebc36ae05a4aeac08c019d7f81d3f4e9e5c256 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2079,6 +2079,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     Iterator<Advancement> advancementIterator();
 
+    // Paper start - Adding getAllBlockDataVariants methods
+    /**
+     * Returns all block data variants of a material block
+     * @param material - The Block material
+     * @return - The List of block data variants
+     */
+    List<BlockData> getAllBlockDataVariants(Material material);
+    // Paper start - Adding getAllBlockDataVariants methods
+
     /**
      * Creates a new {@link BlockData} instance for the specified Material, with
      * all properties initialized to unspecified defaults.
