From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Mon, 27 Nov 2023 19:11:44 +0100
Subject: [PATCH] Adding 3 new builders to BukkitDataFactory


diff --git a/src/main/java/org/bukkit/datafactory/BukkitDataFactory.java b/src/main/java/org/bukkit/datafactory/BukkitDataFactory.java
index 8cbb87204ef9e7c46b2e22274412dd8932e3f3f3..51d448064bb349b211512d4d6c5f18266faf41ef 100644
--- a/src/main/java/org/bukkit/datafactory/BukkitDataFactory.java
+++ b/src/main/java/org/bukkit/datafactory/BukkitDataFactory.java
@@ -3,9 +3,13 @@ package org.bukkit.datafactory;
 import org.bukkit.Keyed;
 import org.bukkit.NamespacedKey;
 import org.bukkit.datafactory.builder.AttributeBuilder;
+import org.bukkit.datafactory.builder.PoiTypeBuilder;
 import org.bukkit.datafactory.builder.VillagerProfessionBuilder;
 
 public interface BukkitDataFactory {
+    AttributeBuilder attributeBuilder();
+    VillagerProfessionBuilder villagerProfessionBuilder();
+    PoiTypeBuilder poiTypeBuilder();
     interface ElementBuilder<B extends Keyed> {
         B build(NamespacedKey namespacedKey);
     }
diff --git a/src/main/java/org/bukkit/datafactory/builder/AttributeBuilder.java b/src/main/java/org/bukkit/datafactory/builder/AttributeBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..83ec448ec6f1186076c26a5949b555cde1973cb3
--- /dev/null
+++ b/src/main/java/org/bukkit/datafactory/builder/AttributeBuilder.java
@@ -0,0 +1,10 @@
+package org.bukkit.datafactory.builder;
+
+import org.bukkit.attribute.Attribute;
+import org.bukkit.datafactory.BukkitDataFactory;
+
+public interface AttributeBuilder extends BukkitDataFactory.ElementBuilder<Attribute> {
+    AttributeBuilder withDefaultValue(double defaultValue);
+    AttributeBuilder withMinValue(double minValue);
+    AttributeBuilder withMaxValue(double maxValue);
+}
diff --git a/src/main/java/org/bukkit/datafactory/builder/PoiTypeBuilder.java b/src/main/java/org/bukkit/datafactory/builder/PoiTypeBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..70f3b3c232c73bda3d2a7359797fe7ac88f3e84d
--- /dev/null
+++ b/src/main/java/org/bukkit/datafactory/builder/PoiTypeBuilder.java
@@ -0,0 +1,13 @@
+package org.bukkit.datafactory.builder;
+
+import org.bukkit.PoiType;
+import org.bukkit.block.data.BlockData;
+import org.bukkit.datafactory.BukkitDataFactory;
+
+import java.util.Set;
+
+public interface PoiTypeBuilder extends BukkitDataFactory.ElementBuilder<PoiType> {
+    PoiTypeBuilder withMatchingStates(Set<BlockData> blockDataSet);
+    PoiTypeBuilder withMaxTickets(int maxTickets);
+    PoiTypeBuilder withValidRange(int validRange);
+}
diff --git a/src/main/java/org/bukkit/datafactory/builder/VillagerProfessionBuilder.java b/src/main/java/org/bukkit/datafactory/builder/VillagerProfessionBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..b3ce764d2febeaed9951a17d10bad3a6f6ae6447
--- /dev/null
+++ b/src/main/java/org/bukkit/datafactory/builder/VillagerProfessionBuilder.java
@@ -0,0 +1,18 @@
+package org.bukkit.datafactory.builder;
+
+import org.bukkit.Material;
+import org.bukkit.NamespacedKey;
+import org.bukkit.PoiType;
+import org.bukkit.Sound;
+import org.bukkit.datafactory.BukkitDataFactory;
+import org.bukkit.entity.Villager;
+
+import java.util.Set;
+
+public interface VillagerProfessionBuilder extends BukkitDataFactory.ElementBuilder<Villager.Profession> {
+     VillagerProfessionBuilder withHeldJobSite(PoiType heldJobSite);
+     VillagerProfessionBuilder withAcquirableJobSite(PoiType acquirableJobSite);
+     VillagerProfessionBuilder withRequestedItems(Set<Material> requestedItems);
+     VillagerProfessionBuilder withSecondaryPoi(Set<Material> secondaryPois);
+    VillagerProfessionBuilder withWorkSound(Sound sound);
+}
