From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Thu, 4 Jan 2024 23:06:02 +0100
Subject: [PATCH] Temp patch 2


diff --git a/src/main/java/de/verdox/mccreativelab/behaviour/ItemBehaviour.java b/src/main/java/de/verdox/mccreativelab/behaviour/ItemBehaviour.java
index 2271581b86b2f7d27984e9b7ff00abf0101efa19..5906e44f6841321fe70e0281ae6ed8a04eaca630 100644
--- a/src/main/java/de/verdox/mccreativelab/behaviour/ItemBehaviour.java
+++ b/src/main/java/de/verdox/mccreativelab/behaviour/ItemBehaviour.java
@@ -3,7 +3,9 @@ package de.verdox.mccreativelab.behaviour;
 import de.verdox.mccreativelab.MultiCustomBehaviour;
 import de.verdox.mccreativelab.recipe.CustomItemData;
 import org.bukkit.World;
+import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
 
 public interface ItemBehaviour {
     MultiCustomBehaviour<CustomItemData, ItemBehaviour> ITEM_BEHAVIOUR = new MultiCustomBehaviour<>("MCCLab - ItemBehaviour");
@@ -11,4 +13,14 @@ public interface ItemBehaviour {
     default BehaviourResult.Void onCraftedBy(World world, Player player, int amount){
         return BehaviourResult.Void.DEFAULT_INSTANCE;
     }
+
+    default BehaviourResult.Object<ItemStack> finishUsingItem(LivingEntity livingEntity, ItemStack usedItem){
+        return BehaviourResult.Object.DEFAULT_INSTANCE;
+    }
+
+    default BehaviourResult.Object<Integer> getMaxStackSize(ItemStack stack){
+        return BehaviourResult.Object.DEFAULT_INSTANCE;
+    }
+
+
 }
