From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Wed, 1 May 2024 15:00:34 +0000
Subject: [PATCH] Add AdvancementBuilder


diff --git a/src/main/java/de/verdox/mccreativelab/CreativeLab.java b/src/main/java/de/verdox/mccreativelab/CreativeLab.java
index f7f6f30e99a78e456aed0be0aeb6d86ee62ae855..a7992255aaed5beb35736447d56b7b2735ed02cb 100644
--- a/src/main/java/de/verdox/mccreativelab/CreativeLab.java
+++ b/src/main/java/de/verdox/mccreativelab/CreativeLab.java
@@ -3,4 +3,5 @@ package de.verdox.mccreativelab;
 public interface CreativeLab {
     de.verdox.mccreativelab.data.DataPackInterceptor getDataPackInterceptor();
     de.verdox.mccreativelab.data.VanillaRegistryManipulator getRegistryManipulator();
+    de.verdox.mccreativelab.advancement.AdvancementBuilder createAdvancement();
 }
diff --git a/src/main/java/de/verdox/mccreativelab/MCCreativeLab.java b/src/main/java/de/verdox/mccreativelab/MCCreativeLab.java
index 1fd9786c6e3635ef6c0a2e5cde08d700ec79a43f..1e301142a2e6341cef550cfd1e4ed1c5e5cd7e40 100644
--- a/src/main/java/de/verdox/mccreativelab/MCCreativeLab.java
+++ b/src/main/java/de/verdox/mccreativelab/MCCreativeLab.java
@@ -15,4 +15,8 @@ public final class MCCreativeLab{
     public static de.verdox.mccreativelab.data.VanillaRegistryManipulator getRegistryManipulator(){
         return creativeLab.getRegistryManipulator();
     }
+
+    public static de.verdox.mccreativelab.advancement.AdvancementBuilder createAdvancement(){
+        return creativeLab.createAdvancement();
+    }
 }
diff --git a/src/main/java/de/verdox/mccreativelab/advancement/AdvancementBuilder.java b/src/main/java/de/verdox/mccreativelab/advancement/AdvancementBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..feae00ebbe2fb0342209d879c562477e2b0a0d5f
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/advancement/AdvancementBuilder.java
@@ -0,0 +1,14 @@
+package de.verdox.mccreativelab.advancement;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.advancement.Advancement;
+
+import java.util.function.Consumer;
+
+public interface AdvancementBuilder {
+    AdvancementBuilder withParent(NamespacedKey namespacedKey);
+    AdvancementBuilder withDisplay(Consumer<AdvancementDisplayBuilder> craftAdvancementDisplayBuilder);
+    AdvancementBuilder withRewards(Consumer<AdvancementRewardBuilder> craftAdvancementRewardBuilder);
+    AdvancementBuilder withRequirements(Consumer<AdvancementRequirementsBuilder> craftAdvancementRewardBuilder);
+    Advancement addToBukkit(NamespacedKey namespacedKey);
+}
diff --git a/src/main/java/de/verdox/mccreativelab/advancement/AdvancementDisplayBuilder.java b/src/main/java/de/verdox/mccreativelab/advancement/AdvancementDisplayBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..8b87db5c84fa134ce41f7131c9023893ec5b126e
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/advancement/AdvancementDisplayBuilder.java
@@ -0,0 +1,28 @@
+package de.verdox.mccreativelab.advancement;
+
+import io.papermc.paper.advancement.AdvancementDisplay;
+import net.kyori.adventure.text.Component;
+import org.bukkit.NamespacedKey;
+import org.bukkit.inventory.ItemStack;
+
+public interface AdvancementDisplayBuilder {
+    AdvancementDisplayBuilder setFrame(AdvancementDisplay.Frame frame);
+
+    AdvancementDisplayBuilder setTitle(Component title);
+
+    AdvancementDisplayBuilder setDescription(Component description);
+
+    AdvancementDisplayBuilder setIcon(ItemStack icon);
+
+    AdvancementDisplayBuilder setShowToast(boolean showToast);
+
+    AdvancementDisplayBuilder setAnnounceToChat(boolean announceToChat);
+
+    AdvancementDisplayBuilder setHidden(boolean hidden);
+
+    AdvancementDisplayBuilder setBackground(NamespacedKey background);
+
+    AdvancementDisplayBuilder setX(float x);
+
+    AdvancementDisplayBuilder setY(float y);
+}
diff --git a/src/main/java/de/verdox/mccreativelab/advancement/AdvancementRequirementsBuilder.java b/src/main/java/de/verdox/mccreativelab/advancement/AdvancementRequirementsBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..e2cbce773b0e225ba927f00004f7b0cfc1804809
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/advancement/AdvancementRequirementsBuilder.java
@@ -0,0 +1,13 @@
+package de.verdox.mccreativelab.advancement;
+
+public interface AdvancementRequirementsBuilder {
+
+    AdvancementRequirementsBuilder withStrategy(Strategy strategy);
+    AdvancementRequirementsBuilder addCriterion(String name);
+
+
+    enum Strategy {
+        OR,
+        AND,
+    }
+}
diff --git a/src/main/java/de/verdox/mccreativelab/advancement/AdvancementRewardBuilder.java b/src/main/java/de/verdox/mccreativelab/advancement/AdvancementRewardBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..39a46d923e38a10a8c822aa4664005bf6fda7530
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/advancement/AdvancementRewardBuilder.java
@@ -0,0 +1,11 @@
+package de.verdox.mccreativelab.advancement;
+
+import org.bukkit.NamespacedKey;
+
+public interface AdvancementRewardBuilder {
+    AdvancementRewardBuilder setExperience(int experience);
+
+    AdvancementRewardBuilder withLootTable(NamespacedKey namespacedKey);
+
+    AdvancementRewardBuilder withRecipe(NamespacedKey namespacedKey);
+}
