From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Thu, 18 Jan 2024 16:18:20 +0100
Subject: [PATCH] Add BlockDestroyCallback to BlockBehaviour


diff --git a/src/main/java/de/verdox/mccreativelab/behaviour/BlockBehaviour.java b/src/main/java/de/verdox/mccreativelab/behaviour/BlockBehaviour.java
index db69e6d95c2537309b62d1a898df2c5b11b003d1..9f82fed2bfdc992f4793934622e1f4784af5f886 100644
--- a/src/main/java/de/verdox/mccreativelab/behaviour/BlockBehaviour.java
+++ b/src/main/java/de/verdox/mccreativelab/behaviour/BlockBehaviour.java
@@ -148,6 +148,18 @@ public interface BlockBehaviour {
         return BehaviourResult.Callback.DEFAULT_INSTANCE;
     }
 
+    /**
+     *
+     * @param location - The block location
+     * @param drop - True if the block dropped items
+     * @param destroyingEntity - The Entity that destroyed the block
+     * @param maxUpdateDepth - The chunk update depth
+     * @return a callback object
+     */
+    default BehaviourResult.Callback onDestroy(Location location, boolean drop, @javax.annotation.Nullable Entity destroyingEntity, int maxUpdateDepth) {
+        return BehaviourResult.Callback.DEFAULT_INSTANCE;
+    }
+
     /**
      * This callback is run after a player interacted with this block. This method is not run on blocks that call bukkit events! Only on those that do not define any specific onUse Behaviour
      *
