From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Thu, 11 Jan 2024 17:27:21 +0100
Subject: [PATCH] Temp patch 2


diff --git a/src/main/java/de/verdox/mccreativelab/ai/EntityBrain.java b/src/main/java/de/verdox/mccreativelab/ai/EntityBrain.java
index 98eeba9031f61892ae8d2a6042c2739fe9bb2bf2..3f4cd909986d20d9d544a2f469602e8eba394034 100644
--- a/src/main/java/de/verdox/mccreativelab/ai/EntityBrain.java
+++ b/src/main/java/de/verdox/mccreativelab/ai/EntityBrain.java
@@ -39,7 +39,7 @@ public interface EntityBrain<E extends LivingEntity> {
     ActivityBuilder<E> createActivityBuilder(EntityActivity entityActivity);
 
     interface ActivityBuilder<E extends LivingEntity>{
-        ActivityBuilder<E> withBehaviour(int priority, PaperAIBehaviour<E> aiBehaviour);
+        ActivityBuilder<E> withBehaviour(int priority, AIBehaviour<E> aiBehaviour);
         ActivityBuilder<E> withRequiredMemory(MemoryKey<?> requiredMemoryKey, MemoryStatus memoryStatus);
         ActivityBuilder<E> withForgettingMemoriesWhenStopped(MemoryKey<?> forgettingMemoryKey);
     }
