From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Fri, 26 Jan 2024 15:30:34 +0100
Subject: [PATCH] Remove wrong @Nullable


diff --git a/src/main/java/de/verdox/mccreativelab/CustomBehaviour.java b/src/main/java/de/verdox/mccreativelab/CustomBehaviour.java
index f43f9487e6db814a47008be029aebb501c999c1f..0ab34266d4cd7c559892517b76673bfa0f16e088 100644
--- a/src/main/java/de/verdox/mccreativelab/CustomBehaviour.java
+++ b/src/main/java/de/verdox/mccreativelab/CustomBehaviour.java
@@ -33,7 +33,7 @@ public class CustomBehaviour<T> {
         this.proxyInterface.addImplementation(behaviour);
     }
 
-    @Nullable
+
     @ApiStatus.Internal
     public T getBehaviour() {
         return proxyInterface.getImplementation();
