From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Wed, 29 May 2024 15:23:38 +0000
Subject: [PATCH] Temp Patch


diff --git a/src/main/java/de/verdox/mccreativelab/CreativeLab.java b/src/main/java/de/verdox/mccreativelab/CreativeLab.java
index a7992255aaed5beb35736447d56b7b2735ed02cb..6ae6fa4dd9ac10c12eeb57310b2fcf434d6229df 100644
--- a/src/main/java/de/verdox/mccreativelab/CreativeLab.java
+++ b/src/main/java/de/verdox/mccreativelab/CreativeLab.java
@@ -1,7 +1,31 @@
 package de.verdox.mccreativelab;
 
+import de.verdox.mccreativelab.container.CustomInventory;
+import org.bukkit.inventory.Inventory;
+import org.jetbrains.annotations.NotNull;
+
 public interface CreativeLab {
     de.verdox.mccreativelab.data.DataPackInterceptor getDataPackInterceptor();
+
     de.verdox.mccreativelab.data.VanillaRegistryManipulator getRegistryManipulator();
+
     de.verdox.mccreativelab.advancement.AdvancementBuilder createAdvancement();
+
+    Inventory openCustomContainerMenu(@NotNull CustomInventory customInventory, @NotNull org.bukkit.entity.Player player, @NotNull net.kyori.adventure.text.Component title);
+
+    /**
+     * Returns a stream of all possible {@link org.bukkit.block.data.BlockData} definitions of a valid block {@link org.bukkit.Material}
+     *
+     * @param blockMaterial - The material of the block
+     * @return - the stream
+     */
+    java.util.stream.Stream<org.bukkit.block.data.BlockData> streamAllBlockDataVariants(org.bukkit.Material blockMaterial);
+
+    /**
+     * Returns an iterator that iterates through all possible {@link org.bukkit.block.data.BlockData} definitions of a valid block {@link org.bukkit.Material}
+     *
+     * @param blockMaterial - The material of the block
+     * @return - the iterator
+     */
+    java.util.Iterator<org.bukkit.block.data.BlockData> iterateThroughAllBlockDataVariants(org.bukkit.Material blockMaterial);
 }
diff --git a/src/main/java/de/verdox/mccreativelab/MCCreativeLab.java b/src/main/java/de/verdox/mccreativelab/MCCreativeLab.java
index 1e301142a2e6341cef550cfd1e4ed1c5e5cd7e40..6c53e6aaf1ff4b09cc7a33128446f55188b0f54e 100644
--- a/src/main/java/de/verdox/mccreativelab/MCCreativeLab.java
+++ b/src/main/java/de/verdox/mccreativelab/MCCreativeLab.java
@@ -1,22 +1,54 @@
 package de.verdox.mccreativelab;
 
-public final class MCCreativeLab{
+import de.verdox.mccreativelab.container.CustomInventory;
+import org.bukkit.block.data.BlockData;
+import org.jetbrains.annotations.NotNull;
+
+public final class MCCreativeLab {
     private static CreativeLab creativeLab;
+
     public static void setCreativeLab(CreativeLab creativeLab) {
-        if(MCCreativeLab.creativeLab != null)
+        if (MCCreativeLab.creativeLab != null)
             throw new UnsupportedOperationException("Cannot redefine singleton CreativeLab");
 
         MCCreativeLab.creativeLab = creativeLab;
     }
-    public static de.verdox.mccreativelab.data.DataPackInterceptor getDataPackInterceptor(){
+
+    public static de.verdox.mccreativelab.data.DataPackInterceptor getDataPackInterceptor() {
         return creativeLab.getDataPackInterceptor();
     }
 
-    public static de.verdox.mccreativelab.data.VanillaRegistryManipulator getRegistryManipulator(){
+    public static de.verdox.mccreativelab.data.VanillaRegistryManipulator getRegistryManipulator() {
         return creativeLab.getRegistryManipulator();
     }
 
-    public static de.verdox.mccreativelab.advancement.AdvancementBuilder createAdvancement(){
+    public static de.verdox.mccreativelab.advancement.AdvancementBuilder createAdvancement() {
         return creativeLab.createAdvancement();
     }
+
+    public static org.bukkit.inventory.Inventory openCustomContainerMenu(@NotNull CustomInventory customInventory, @NotNull org.bukkit.entity.Player player, @NotNull net.kyori.adventure.text.Component title) {
+        return creativeLab.openCustomContainerMenu(customInventory, player, title);
+    }
+
+
+    /**
+     * Returns a stream of all possible {@link org.bukkit.block.data.BlockData} definitions of a valid block {@link org.bukkit.Material}
+     *
+     * @param blockMaterial - The material of the block
+     * @return - the stream
+     */
+    public static java.util.stream.Stream<BlockData> streamAllBlockDataVariants(org.bukkit.Material blockMaterial) {
+        return creativeLab.streamAllBlockDataVariants(blockMaterial);
+    }
+
+    /**
+     * Returns an iterator that iterates through all possible {@link org.bukkit.block.data.BlockData} definitions of a valid block {@link org.bukkit.Material}
+     *
+     * @param blockMaterial - The material of the block
+     * @return - the iterator
+     */
+    public static java.util.Iterator<BlockData> iterateThroughAllBlockDataVariants(org.bukkit.Material blockMaterial) {
+        return creativeLab.iterateThroughAllBlockDataVariants(blockMaterial);
+    }
+
 }
diff --git a/src/main/java/de/verdox/mccreativelab/container/CustomInventory.java b/src/main/java/de/verdox/mccreativelab/container/CustomInventory.java
new file mode 100644
index 0000000000000000000000000000000000000000..7ea085bfa2de6f6396316a4d01fa52499c0644c3
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/container/CustomInventory.java
@@ -0,0 +1,41 @@
+package de.verdox.mccreativelab.container;
+
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+
+public interface CustomInventory {
+
+    Slot[] getSlots();
+    ItemStack quickMoveStack(Player player, int slot);
+
+    interface Slot {
+        default void onQuickCraft(ItemStack stack, int amount) {
+        }
+
+        default void onSwapCraft(int amount) {
+        }
+
+        default void onTake(Player player, ItemStack stack) {
+        }
+
+        default boolean mayPlace(ItemStack stack) {
+            return true;
+        }
+
+        default int getMaxStackSize() {
+            return 64;
+        }
+
+        default int getMaxStackSize(ItemStack stack) {
+            return Math.min(getMaxStackSize(), stack.getMaxStackSize());
+        }
+
+        default boolean isActive() {
+            return true;
+        }
+
+        default boolean mayPickUp(Player player){
+            return true;
+        }
+    }
+}
diff --git a/src/main/java/de/verdox/mccreativelab/data/DataPackInterceptor.java b/src/main/java/de/verdox/mccreativelab/data/DataPackInterceptor.java
index 59a7d6ffa7a1d525c36956c96b3c97800a1cb7e2..221c6b80854e8efd165b554369b5d36d5127f630 100644
--- a/src/main/java/de/verdox/mccreativelab/data/DataPackInterceptor.java
+++ b/src/main/java/de/verdox/mccreativelab/data/DataPackInterceptor.java
@@ -5,6 +5,7 @@ import net.kyori.adventure.key.Key;
 
 import javax.annotation.Nullable;
 import java.util.function.Consumer;
+import java.util.function.Function;
 
 public interface DataPackInterceptor {
     DataPackInterceptor exclude(String pathContains);
@@ -13,7 +14,21 @@ public interface DataPackInterceptor {
 
     DataPackInterceptor exclude(PackAssetType packAssetType, Key key);
 
-    DataPackInterceptor modify(PackAssetType packAssetType, Key key, Consumer<DataPackAsset> consumer);
+    /**
+     * Modifies a data pack element loaded into the server. The provided function modifies the asset and returns true if the asset should be loaded.
+     * @param packAssetType - The Pack Asset Type
+     * @param key - The Key
+     * @param modifier - The modification function
+     * @return - The interceptor
+     */
+    DataPackInterceptor modify(PackAssetType packAssetType, Key key, Function<DataPackAsset, Boolean> modifier);
+
+    /**
+     * Modifies a data pack element loaded into the server. The provided function modifies the asset and returns true if the asset should be loaded.
+     * @param modifier - The modification function
+     * @return - The interceptor
+     */
+    DataPackInterceptor modify(Function<DataPackAsset, Boolean> modifier);
 
     DataPackInterceptor onInstall(Consumer<DataPackAsset> installCallback);
 
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index e07048c269c51ffa943ba2fe0aa2985db0460a17..b409c317dd6c18eb5447f6f6b983da98e8a02ed3 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -2841,7 +2841,7 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     int getSimulationDistance();
 
     /**
-     * Returns the {@link WorldGenData} of this world
+     * Returns the {@link io.papermc.paper.world.worldgen.WorldGenData} of this world
      * @return the world gen data
      */
     public io.papermc.paper.world.worldgen.WorldGenData getWorldGenData();
