From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Tue, 5 Dec 2023 21:21:01 +0100
Subject: [PATCH] Adding MCCreativeLabExtension plugin


diff --git a/build.gradle.kts b/build.gradle.kts
index 763e13b1b1306ecbfb58b600463b2ca900ecef9b..fac324c3864e62fc1bd476e74acd2ed8c37ccdfd 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -193,37 +193,51 @@ tasks.registerRunTask("runServer") {
     description = "Spin up a test server from the Mojang mapped server jar"
     classpath(tasks.includeMappings.flatMap { it.outputJar })
     classpath(runtimeClasspathWithoutVanillaServer)
+    dependsOn(":plugin-extension:build")
+    jvmArgs("-Dpaper.debug-sync-loads=true", "-Dfile.encoding=UTF-8")
 }
 
 tasks.registerRunTask("runReobfServer") {
     description = "Spin up a test server from the reobfJar output jar"
     classpath(tasks.reobfJar.flatMap { it.outputJar })
     classpath(runtimeClasspathWithoutVanillaServer)
+    dependsOn(":plugin-extension:build")
+    jvmArgs("-Dpaper.debug-sync-loads=true", "-Dfile.encoding=UTF-8")
 }
 
 tasks.registerRunTask("runDevServer") {
     description = "Spin up a test server without assembling a jar"
     classpath(sourceSets.main.map { it.runtimeClasspath })
     jvmArgs("-DPaper.pushPaperAssetsRoot=true")
+    dependsOn(":plugin-extension:build")
+    jvmArgs("-Dpaper.debug-sync-loads=true", "-Dfile.encoding=UTF-8")
 }
 
 tasks.registerRunTask("runBundler") {
     description = "Spin up a test server from the Mojang mapped bundler jar"
     classpath(rootProject.tasks.named<io.papermc.paperweight.tasks.CreateBundlerJar>("createMojmapBundlerJar").flatMap { it.outputZip })
     mainClass.set(null as String?)
+    dependsOn(":plugin-extension:build")
+    jvmArgs("-Dpaper.debug-sync-loads=true", "-Dfile.encoding=UTF-8")
 }
 tasks.registerRunTask("runReobfBundler") {
     description = "Spin up a test server from the reobf bundler jar"
     classpath(rootProject.tasks.named<io.papermc.paperweight.tasks.CreateBundlerJar>("createReobfBundlerJar").flatMap { it.outputZip })
     mainClass.set(null as String?)
+    dependsOn(":plugin-extension:build")
+    jvmArgs("-Dpaper.debug-sync-loads=true", "-Dfile.encoding=UTF-8")
 }
 tasks.registerRunTask("runPaperclip") {
     description = "Spin up a test server from the Mojang mapped Paperclip jar"
     classpath(rootProject.tasks.named<io.papermc.paperweight.tasks.CreatePaperclipJar>("createMojmapPaperclipJar").flatMap { it.outputZip })
     mainClass.set(null as String?)
+    dependsOn(":plugin-extension:build")
+    jvmArgs("-Dpaper.debug-sync-loads=true", "-Dfile.encoding=UTF-8")
 }
 tasks.registerRunTask("runReobfPaperclip") {
     description = "Spin up a test server from the reobf Paperclip jar"
     classpath(rootProject.tasks.named<io.papermc.paperweight.tasks.CreatePaperclipJar>("createReobfPaperclipJar").flatMap { it.outputZip })
     mainClass.set(null as String?)
+    dependsOn(":plugin-extension:build")
+    jvmArgs("-Dpaper.debug-sync-loads=true", "-Dfile.encoding=UTF-8")
 }
