From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Sun, 12 Nov 2023 11:14:37 +0100
Subject: [PATCH] TestPatch


diff --git a/src/main/java/de/verdox/mccreativelab/TestClass.java b/src/main/java/de/verdox/mccreativelab/TestClass.java
new file mode 100644
index 0000000000000000000000000000000000000000..bf98160c04a479aa05162d2b680de73bdcee777d
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/TestClass.java
@@ -0,0 +1,5 @@
+package de.verdox.mccreativelab;
+
+public class TestClass {
+    public int bla = 1;
+}
