From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Fri, 9 Jun 2023 22:07:24 +0200
Subject: [PATCH] Ability to open an inventory with a different title


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index c3dbcb317b7d366feb31f707ad1199c60169f07f..51d6d12cacea21e4fa42759c531346dfc80dbc1e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -311,6 +311,66 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         return this.getHandle().containerMenu.getBukkitView();
     }
 
+    // MCCLab start - Open Inventory with different title
+    @Override
+    public @org.jetbrains.annotations.Nullable InventoryView openInventory(@org.jetbrains.annotations.NotNull Inventory inventory, net.kyori.adventure.text.Component title) {
+        if (!(this.getHandle() instanceof ServerPlayer)) return null;
+        ServerPlayer player = (ServerPlayer) this.getHandle();
+        AbstractContainerMenu formerContainer = this.getHandle().containerMenu;
+
+        MenuProvider iinventory = null;
+        if (inventory instanceof CraftInventoryDoubleChest) {
+            iinventory = ((CraftInventoryDoubleChest) inventory).tile;
+        } else if (inventory instanceof CraftInventoryLectern) {
+            iinventory = ((CraftInventoryLectern) inventory).tile;
+        } else if (inventory instanceof CraftInventory) {
+            CraftInventory craft = (CraftInventory) inventory;
+            if (craft.getInventory() instanceof MenuProvider) {
+                iinventory = (MenuProvider) craft.getInventory();
+            }
+        }
+
+        if (iinventory instanceof MenuProvider) {
+            if (iinventory instanceof BlockEntity) {
+                BlockEntity te = (BlockEntity) iinventory;
+                if (!te.hasLevel()) {
+                    te.setLevel(this.getHandle().level());
+                }
+            }
+        }
+
+        MenuType<?> container = CraftContainer.getNotchInventoryType(inventory);
+        if (iinventory instanceof MenuProvider) {
+            this.getHandle().openMenu(iinventory);
+        } else {
+            CraftHumanEntity.openCustomInventory(inventory, player, container, title);
+        }
+
+        if (this.getHandle().containerMenu == formerContainer) {
+            return null;
+        }
+        this.getHandle().containerMenu.checkReachable = false;
+        return this.getHandle().containerMenu.getBukkitView();
+    }
+
+    private static void openCustomInventory(Inventory inventory, ServerPlayer player, MenuType<?> windowType, net.kyori.adventure.text.Component title) {
+        if (player.connection == null) return;
+        Preconditions.checkArgument(windowType != null, "Unknown windowType");
+        AbstractContainerMenu container = new CraftContainer(inventory, player, player.nextContainerCounter());
+
+        container = CraftEventFactory.callInventoryOpenEvent(player, container);
+        if (container == null) return;
+
+        //player.connection.send(new ClientboundOpenScreenPacket(container.containerId, windowType, CraftChatMessage.fromString(title)[0])); // Paper - comment
+        if (!player.isImmobile())
+            player.connection.send(new ClientboundOpenScreenPacket(container.containerId, windowType, io.papermc.paper.adventure.PaperAdventure.asVanilla(title))); // Paper
+        player.containerMenu = container;
+        player.initMenu(container);
+    }
+
+    // MCCLab end - Open Inventory with different title
+
+
     @Override
     public InventoryView openInventory(Inventory inventory) {
         if (!(this.getHandle() instanceof ServerPlayer)) return null;
