From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Mon, 13 Nov 2023 21:26:23 +0100
Subject: [PATCH] Adding alwaysTick flag for Chunks, so they can tick with no
 nearby player.


diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index fa170cc1ce7011d201295b89718292d696c7fc24..ad207ccab1fc64b540bf5cb05b3d78c5e71849e7 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -1215,4 +1215,19 @@ public class LevelChunk extends ChunkAccess {
             return "Level ticker for " + s + "@" + this.getPos();
         }
     }
+
+    // MCCLab - start
+
+    private boolean alwaysTicking;
+
+    public void setAlwaysTicking(boolean alwaysTicking) {
+        this.alwaysTicking = alwaysTicking;
+    }
+
+    public boolean isAlwaysTicking() {
+        return alwaysTicking;
+    }
+
+    // MCCLab - end
+
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index b6716f2b0037d12fb9956d7d1d0fab9b86cf4433..9573c7c93159bc6253660caf9afc7b1ce1583890 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -405,4 +405,18 @@ public class CraftChunk implements Chunk {
     static {
         Arrays.fill(FULL_LIGHT, (byte) 0xFF);
     }
+
+    // MCLab - start
+
+    @Override
+    public void setAlwaysTicking(boolean alwaysTicking) {
+        worldServer.getChunk(getX(), getZ()).setAlwaysTicking(alwaysTicking);
+    }
+
+    @Override
+    public boolean isAlwaysTicking() {
+        return worldServer.getChunk(getX(), getZ()).isAlwaysTicking();
+    }
+
+    // MCLab - end
 }
