From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Fri, 17 Nov 2023 17:05:03 +0100
Subject: [PATCH] Adding isRainingAt method to World.java


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 9af8c377e140923f84fd4d0eaa1829ac782f8e73..c24f6bbdff9bde6cbf863149ce7cc0175a52a10b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -166,6 +166,10 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     // MCCLab - start
     @Override
+    public boolean isRainingAt(Location location) {
+        return world.isRainingAt(new BlockPos(location.getBlockX(),location.getBlockY(),location.getBlockZ()));
+    }
+    @Override
     public boolean growTreeIfSapling(@NotNull Location location) {
         var blockPos = new BlockPos(location.getBlockX(),location.getBlockY(),location.getBlockZ());
         var block = world.getBlockIfLoaded(blockPos);
