From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Fri, 17 Nov 2023 17:05:03 +0100
Subject: [PATCH] Adding isRainingAt method to World.java


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index f51087a8125a87db1668b31e024c07e99c0568f6..b842519f6595ee7c0cc21621be17502915084146 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -166,6 +166,10 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     // MCCLab - start
     @Override
+    public boolean isRainingAt(Location location) {
+        return world.isRainingAt(new BlockPos(location.getBlockX(),location.getBlockY(),location.getBlockZ()));
+    }
+    @Override
     public boolean growTreeIfSapling(@NotNull Location location) {
         var blockPos = new BlockPos(location.getBlockX(),location.getBlockY(),location.getBlockZ());
         var block = world.getBlockIfLoaded(blockPos);
