From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Fri, 17 Nov 2023 17:05:03 +0100
Subject: [PATCH] Adding isRainingAt method to World.java


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index d7f7e01c0f95f6a23c577cfeea678b18015a16f2..1f1c7046d81f9b71cd5e8be04726b9d05c0f2204 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -160,6 +160,10 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     // MCCLab - start
     @Override
+    public boolean isRainingAt(Location location) {
+        return world.isRainingAt(new BlockPos(location.getBlockX(),location.getBlockY(),location.getBlockZ()));
+    }
+    @Override
     public boolean growTreeIfSapling(@NotNull Location location) {
         var blockPos = new BlockPos(location.getBlockX(),location.getBlockY(),location.getBlockZ());
         var block = world.getBlockIfLoaded(blockPos);
