From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tim Southwick <NickNackGus@gmail.com>
Date: Sun, 19 Apr 2020 19:34:47 -0400
Subject: [PATCH] Monumenta - Selectors require targets to be alive


diff --git a/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java b/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
index 466d1d8d80df028ff842ce21be198be6a1c77b42..51c7854c6794fc9232cf788e532104415fe51f97 100644
--- a/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
+++ b/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
@@ -125,9 +125,7 @@ public class EntitySelectorParser {
         this.level = MinMaxBounds.Ints.ANY;
         this.rotX = WrappedMinMaxBounds.ANY;
         this.rotY = WrappedMinMaxBounds.ANY;
-        this.predicate = (entity) -> {
-            return true;
-        };
+        this.predicate = Entity::isAlive;
         this.order = EntitySelectorParser.ORDER_ARBITRARY;
         this.suggestions = EntitySelectorParser.SUGGEST_NOTHING;
         this.reader = reader;
