From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tim Southwick <NickNackGus@gmail.com>
Date: Sun, 19 Apr 2020 19:34:47 -0400
Subject: [PATCH] Monumenta - Selectors require targets to be alive


diff --git a/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java b/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
index 015d01242a9e8e7c6ef5b6bbf1b6d6ad0c8f36ca..8945a14b455f6c31a55d78e4951239b80e225f8e 100644
--- a/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
+++ b/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
@@ -132,9 +132,7 @@ public class EntitySelectorParser {
         this.level = MinMaxBounds.Ints.ANY;
         this.rotX = WrappedMinMaxBounds.ANY;
         this.rotY = WrappedMinMaxBounds.ANY;
-        this.predicate = (entity) -> {
-            return true;
-        };
+        this.predicate = Entity::isAlive;
         this.order = EntitySelectorParser.ORDER_ARBITRARY;
         this.suggestions = EntitySelectorParser.SUGGEST_NOTHING;
         this.reader = reader;
