From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Fri, 13 Mar 2020 17:33:42 -0400
Subject: [PATCH] Monumenta - Make armor unbreaking work the same as tools

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 66f808cabcf6a9a6584849b285f1c60133adc7b4..f79f39a5eb38925ac252220ba9fa04255ce6b466 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -535,6 +535,12 @@ public final class ItemStack {
                 j = EnchantmentHelper.getItemEnchantmentLevel(Enchantments.UNBREAKING, this);
                 int k = 0;
 
+                /* MONUMENTA:
+                 * Reduce giant spikes of gear damage to make high-damage mobs less destructive to gear
+                 * https://www.wolframalpha.com/input/?i=plot+min%28x%2F8%2B2%2Cx%2F4%29+and+min%28sqrt%28x%29%2Cx%2F4%29+and+x%2F4%2C+x%3D1..40
+                 */
+                amount = (int)Math.min(amount, Math.sqrt(amount * 4));
+
                 for (int l = 0; j > 0 && l < amount; ++l) {
                     if (DigDurabilityEnchantment.shouldIgnoreDurabilityDrop(this, j, random)) {
                         ++k;
diff --git a/src/main/java/net/minecraft/world/item/enchantment/DigDurabilityEnchantment.java b/src/main/java/net/minecraft/world/item/enchantment/DigDurabilityEnchantment.java
index f8ccc90054fd29beeafbc92598dc16e733a63333..051c604e331c310d0761260fe7825e7549ffbff4 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/DigDurabilityEnchantment.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/DigDurabilityEnchantment.java
@@ -31,10 +31,7 @@ public class DigDurabilityEnchantment extends Enchantment {
     }
 
     public static boolean shouldIgnoreDurabilityDrop(ItemStack item, int level, Random random) {
-        if (item.getItem() instanceof ArmorItem && random.nextFloat() < 0.6F) {
-            return false;
-        } else {
-            return random.nextInt(level + 1) > 0;
-        }
+        /* MONUMENTA: All items work like tools for unbreaking */
+        return random.nextInt(level + 1) > 0;
     }
 }
