From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Fri, 6 Sep 2019 20:36:55 -0400
Subject: [PATCH] Monumenta - DamageCause.CUSTOM support for class abilities

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index aaf47d3f58819ead3553973c159c5c6eec4abe43..4e879be70fe05e00623922bc8e6acb0fbf8c3cf6 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -985,6 +985,8 @@ public class CraftEventFactory {
                 } else if (damager.getBukkitEntity() instanceof Projectile) {
                     cause = DamageCause.PROJECTILE;
                 }
+            } else if ("custom".equals(source.msgId)) {
+                cause = DamageCause.CUSTOM;
             } else if ("thorns".equals(source.msgId)) {
                 cause = DamageCause.THORNS;
             }
