From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Wed, 29 May 2024 12:52:40 +0000
Subject: [PATCH] Fix compilation error in Brain.java


diff --git a/src/main/java/net/minecraft/world/entity/ai/Brain.java b/src/main/java/net/minecraft/world/entity/ai/Brain.java
index c44a88cc8143850a918a2b56dd8341f6f4eeba03..69abf01ae5a3b64868fa36367967cb679b3a9506 100644
--- a/src/main/java/net/minecraft/world/entity/ai/Brain.java
+++ b/src/main/java/net/minecraft/world/entity/ai/Brain.java
@@ -98,7 +98,7 @@ public class Brain<E extends LivingEntity> {
                             .resultOrPartial(Brain.LOGGER::error)
                             .map(Builder::build)
                             .orElseGet(ImmutableList::of);
-                        return DataResult.success(new Brain<>(memoryModules, sensors, immutableList, mutableObject::getValue));
+                        return DataResult.success(new Brain(memoryModules, sensors, immutableList, mutableObject::getValue));
                     }
 
                     private <T, U> DataResult<Brain.MemoryValue<U>> captureRead(MemoryModuleType<U> memoryType, DynamicOps<T> ops, T value) {
