From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sun, 3 Dec 2023 14:20:30 +0100
Subject: [PATCH] Implementing Basic Item Format


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 58c38bc4361ddf24716f326b0c6fc626d434756e..82ca17b8bdbee786a02206c7c045983ffc933ef8 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -306,11 +306,13 @@ public final class ItemStack implements DataComponentHolder {
         this.count = count;
         this.components = components;
         this.getItem().verifyComponentsAfterLoad(this);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     private ItemStack(@Nullable Void v) {
         this.item = null;
         this.components = new PatchedDataComponentMap(DataComponentMap.EMPTY);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     public static DataResult<Unit> validateComponents(DataComponentMap components) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 9f74a685e0ec51e66f2900d4782849438eb56320..5596f1b3733d19c614c85b18f94713b465e0e7f9 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -146,6 +146,7 @@ public final class CraftItemStack extends ItemStack {
         this.setAmount(amount);
         this.setDurability(durability);
         this.setItemMeta(itemMeta);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(this);
     }
 
     @Override
