From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Wed, 29 Apr 2020 21:17:15 +0000
Subject: [PATCH] Monumenta - Reset last player hurt time on taking any damage

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 572f9ca81b78c6229725f6693940ac0a70ecdfd5..5f3cea1a6c2df02dd6241ce2f78c66fbf3298ac6 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1370,6 +1370,10 @@ public abstract class LivingEntity extends Entity {
             this.hurtDir = 0.0F;
             Entity entity1 = source.getEntity();
 
+            if (this.lastHurtByPlayer != null) {
+                this.lastHurtByPlayerTime = 100;
+            }
+
             if (entity1 != null) {
                 if (entity1 instanceof LivingEntity && !source.isNoAggro()) {
                     this.setLastHurtByMob((LivingEntity) entity1);
