From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sun, 3 Dec 2023 14:20:30 +0100
Subject: [PATCH] Implementing Basic Item Format


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 6f530c3b7b38cf62bb455d55931515b82e31088d..b1674100cf808ea97928c150619ff1e37ae5cba1 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -307,11 +307,13 @@ public final class ItemStack implements DataComponentHolder {
         this.count = count;
         this.components = components;
         this.getItem().verifyComponentsAfterLoad(this);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     private ItemStack(@Nullable Void v) {
         this.item = null;
         this.components = new PatchedDataComponentMap(DataComponentMap.EMPTY);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     public static DataResult<Unit> validateComponents(DataComponentMap components) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 546c3cddc5feafccdaf552776744655186b62bce..da5a66cc2cb2d464add89d3d2cdecf62b792000d 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -147,6 +147,7 @@ public final class CraftItemStack extends ItemStack {
         this.setAmount(amount);
         this.setDurability(durability);
         this.setItemMeta(itemMeta);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(this);
     }
 
     @Override
