From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Thu, 21 Mar 2024 17:45:35 +0100
Subject: [PATCH] Adding PersistentDataObjects


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 696d075ca2883f3c37e35f983c4d020e5db89d16..f4b5c0b7fd1cb1bd702feb6e1d7f86a240a432ab 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1035,6 +1035,13 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         if (this.metricsRecorder.isRecording()) {
             this.cancelRecordingMetrics();
         }
+        // Paper start - Adding PersistentDataObjects
+        long start = System.currentTimeMillis();
+        int saved = org.bukkit.craftbukkit.persistence.CraftPersistentDataObjectCache.saveAllPersistentDataObjects();
+        long end = System.currentTimeMillis() - start;
+        if(saved > 0)
+            MinecraftServer.LOGGER.info("Took "+end+" ms to save "+saved+" persistent data objects");
+        // Paper end - Adding PersistentDataObjects
 
         MinecraftServer.LOGGER.info("Stopping server");
         Commands.COMMAND_SENDING_POOL.shutdownNow(); // Paper - Perf: Async command map building; Shutdown and don't bother finishing
diff --git a/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
index 9d867f1659433ea15f281c8b441db7e339013100..3a605108e69f0db4792bfdb8a262a2689104d3ee 100644
--- a/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
+++ b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
@@ -20,6 +20,7 @@ public class CraftPersistentDataContainer extends io.papermc.paper.persistence.P
 
     private final Map<String, Tag> customDataTags = new HashMap<>();
     // Paper - move to PersistentDataContainerView
+    private final CraftPersistentDataObjectCache craftPersistentDataObjectCache; // Paper - Adding PersistentDataObjects
 
     public CraftPersistentDataContainer(Map<String, Tag> customTags, CraftPersistentDataTypeRegistry registry) {
         this(registry);
@@ -28,6 +29,7 @@ public class CraftPersistentDataContainer extends io.papermc.paper.persistence.P
 
     public CraftPersistentDataContainer(CraftPersistentDataTypeRegistry registry) {
         super(registry); // Paper - move to PersistentDataContainerView
+        this.craftPersistentDataObjectCache = new CraftPersistentDataObjectCache(this, registry, customDataTags); // Paper - Adding PersistentDataObjects
     }
 
     // Paper start
@@ -44,6 +46,10 @@ public class CraftPersistentDataContainer extends io.papermc.paper.persistence.P
         Preconditions.checkArgument(value != null, "The provided value cannot be null");
 
         this.customDataTags.put(key.toString(), this.registry.wrap(type, type.toPrimitive(value, this.adapterContext)));
+        // Paper start - Adding PersistentDataObjects
+        // Remove persistent data object if the key was taken by another entry
+        craftPersistentDataObjectCache.persistentDataObjectMap.remove(key);
+        // Paper end - Adding PersistentDataObjects
     }
 
     // Paper - move to PersistentDataContainerView
@@ -52,6 +58,7 @@ public class CraftPersistentDataContainer extends io.papermc.paper.persistence.P
     @Override
     public Set<NamespacedKey> getKeys() {
         Set<NamespacedKey> keys = new HashSet<>();
+        craftPersistentDataObjectCache.saveObjectsToTags(true); // Paper - Adding PersistentDataObjects
 
         this.customDataTags.keySet().forEach(key -> {
             String[] keyData = key.split(":", 2);
@@ -67,24 +74,29 @@ public class CraftPersistentDataContainer extends io.papermc.paper.persistence.P
     public void remove(@NotNull NamespacedKey key) {
         Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
 
+        this.craftPersistentDataObjectCache.removePersistentDataObject(key, false); // Paper - Adding PersistentDataObjects
         this.customDataTags.remove(key.toString());
     }
 
     @Override
     public boolean isEmpty() {
-        return this.customDataTags.isEmpty();
+        return this.customDataTags.isEmpty() && craftPersistentDataObjectCache.persistentDataObjectMap.isEmpty(); // Paper - Adding PersistentDataObjects
     }
 
     @NotNull
     @Override
     public void copyTo(PersistentDataContainer other, boolean replace) {
         Preconditions.checkArgument(other != null, "The target container cannot be null");
+        craftPersistentDataObjectCache.saveObjectsToTags(true); // Paper - Adding PersistentDataObjects
 
         CraftPersistentDataContainer target = (CraftPersistentDataContainer) other;
+        target.craftPersistentDataObjectCache.saveObjectsToTags(true); // Paper - Adding PersistentDataObjects
         if (replace) {
             target.customDataTags.putAll(this.customDataTags);
+            target.craftPersistentDataObjectCache.persistentDataObjectMap.putAll(this.craftPersistentDataObjectCache.persistentDataObjectMap); // Paper - Adding PersistentDataObjects
         } else {
             this.customDataTags.forEach(target.customDataTags::putIfAbsent);
+            this.craftPersistentDataObjectCache.persistentDataObjectMap.forEach(target.craftPersistentDataObjectCache.persistentDataObjectMap::putIfAbsent); // Paper - Adding PersistentDataObjects
         }
     }
 
@@ -102,11 +114,12 @@ public class CraftPersistentDataContainer extends io.papermc.paper.persistence.P
         Map<String, Tag> myRawMap = this.getRaw();
         Map<String, Tag> theirRawMap = ((CraftPersistentDataContainer) obj).getRaw();
 
-        return Objects.equals(myRawMap, theirRawMap);
+        return Objects.equals(myRawMap, theirRawMap) && this.craftPersistentDataObjectCache.persistentDataObjectMap.equals(((CraftPersistentDataContainer) obj).craftPersistentDataObjectCache.persistentDataObjectMap); // Paper - Adding PersistentDataObjects
     }
 
     public CompoundTag toTagCompound() {
         CompoundTag tag = new CompoundTag();
+        craftPersistentDataObjectCache.saveObjectsToTags(true); // Paper - Adding PersistentDataObjects
         for (Entry<String, Tag> entry : this.customDataTags.entrySet()) {
             tag.put(entry.getKey(), entry.getValue());
         }
@@ -148,6 +161,7 @@ public class CraftPersistentDataContainer extends io.papermc.paper.persistence.P
 
     // Paper start
     public void clear() {
+        this.craftPersistentDataObjectCache.persistentDataObjectMap.clear(); // Paper - Adding PersistentDataObjects
         this.customDataTags.clear();
     }
     // Paper end
@@ -168,9 +182,36 @@ public class CraftPersistentDataContainer extends io.papermc.paper.persistence.P
 
     // Paper start - deep clone tags
     public Map<String, Tag> getTagsCloned() {
+        craftPersistentDataObjectCache.saveObjectsToTags(false); // Paper - Adding PersistentDataObjects
         final Map<String, Tag> tags = new HashMap<>();
         this.customDataTags.forEach((key, tag) -> tags.put(key, tag.copy()));
         return tags;
     }
     // Paper end - deep clone tags
+
+    // Paper start - Adding PersistentDataObjects
+    @Override
+    public CraftPersistentDataObjectCache getPersistentDataObjectCache() {
+        return craftPersistentDataObjectCache;
+    }
+
+    @Override
+    public boolean has(NamespacedKey key) {
+        craftPersistentDataObjectCache.serializeObjectAtKeyBeforeLookup(key); // Paper - Adding PersistentDataObjects
+        return super.has(key);
+    }
+
+    @Override
+    public <P, C> boolean has(NamespacedKey key, PersistentDataType<P, C> type) {
+        craftPersistentDataObjectCache.serializeObjectAtKeyBeforeLookup(key); // Paper - Adding PersistentDataObjects
+        return super.has(key, type);
+    }
+
+    @Override
+    public <P, C> @org.checkerframework.checker.nullness.qual.Nullable C get(NamespacedKey key, PersistentDataType<P, C> type) {
+        craftPersistentDataObjectCache.serializeObjectAtKeyBeforeLookup(key); // Paper - Adding PersistentDataObjects
+        return super.get(key, type);
+    }
+    // Paper end - Adding PersistentDataObjects
+
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataObjectCache.java b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataObjectCache.java
new file mode 100644
index 0000000000000000000000000000000000000000..69d1eaa764313fef9cc063b6ca2f0f58ae174e73
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataObjectCache.java
@@ -0,0 +1,131 @@
+package org.bukkit.craftbukkit.persistence;
+
+import com.google.common.base.Preconditions;
+import net.minecraft.nbt.Tag;
+import org.bukkit.NamespacedKey;
+import org.bukkit.persistence.PersistentDataContainer;
+import org.bukkit.persistence.PersistentDataObject;
+import org.bukkit.persistence.PersistentDataObjectCache;
+import org.bukkit.persistence.PersistentDataType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Level;
+
+public class CraftPersistentDataObjectCache implements PersistentDataObjectCache {
+    private static final Set<CraftPersistentDataContainer> weakReferences = Collections.newSetFromMap(new WeakHashMap<>());
+
+    public static int saveAllPersistentDataObjects(){
+        synchronized (weakReferences){
+            int counter = 0;
+            for (CraftPersistentDataContainer weakReference : weakReferences) {
+                CraftPersistentDataObjectCache craftPersistentDataObjectCache = weakReference.getPersistentDataObjectCache();
+                craftPersistentDataObjectCache.saveObjectsToTags(false);
+                craftPersistentDataObjectCache.persistentDataObjectMap.clear();
+                counter++;
+            }
+            return counter;
+        }
+    }
+
+    private final CraftPersistentDataContainer parentContainer;
+    private final CraftPersistentDataTypeRegistry registry;
+    private final Map<String, Tag> tags;
+    final Map<NamespacedKey, PersistentDataObject> persistentDataObjectMap = new ConcurrentHashMap<>();
+    CraftPersistentDataObjectCache(CraftPersistentDataContainer parentContainer, CraftPersistentDataTypeRegistry registry, Map<String, Tag> tags){
+        this.parentContainer = parentContainer;
+        this.registry = registry;
+        this.tags = tags;
+    }
+    @Override
+    public <T extends PersistentDataObject> @NotNull T loadOrSupplyPersistentDataObject(@NotNull NamespacedKey key, java.util.function.Supplier<T> newObjectSupplier) {
+        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
+        Preconditions.checkArgument(newObjectSupplier != null, "The provided newObjectSupplier cannot be null");
+        if (persistentDataObjectMap.containsKey(key))
+            return (T) persistentDataObjectMap.get(key);
+
+        T newObject = newObjectSupplier.get();
+        Preconditions.checkArgument(newObject != null, "The newObject supplied cannot be null");
+
+        readObjectFromTags(key, newObject);
+        persistentDataObjectMap.put(key, newObject);
+        markParentAsDirty();
+        synchronized (weakReferences){
+            weakReferences.add(parentContainer);
+        }
+        return newObject;
+    }
+
+    @Override
+    public <T extends PersistentDataObject> @Nullable T loadPersistentDataObject(@NotNull NamespacedKey key, @NotNull Class<? extends T> type) {
+        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
+        Preconditions.checkArgument(type != null, "The provided type cannot be null");
+        if (persistentDataObjectMap.containsKey(key))
+            return (T) persistentDataObjectMap.get(key);
+        return null;
+    }
+
+    @Override
+    public void removePersistentDataObject(@NotNull NamespacedKey key, boolean serializeBeforeRemoval) {
+        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
+        org.bukkit.persistence.PersistentDataObject oldValue = persistentDataObjectMap.remove(key);
+        if(serializeBeforeRemoval && oldValue != null)
+            saveObjectToTags(key, oldValue, true);
+        else {
+            markParentAsDirty();
+        }
+        if(persistentDataObjectMap.isEmpty())
+            synchronized (weakReferences){
+                weakReferences.remove(this);
+            }
+    }
+
+    void saveObjectsToTags(boolean removeWhenEmpty) {
+        for (NamespacedKey namespacedKey : Set.copyOf(persistentDataObjectMap.keySet())) {
+            org.bukkit.persistence.PersistentDataObject persistentDataObject = persistentDataObjectMap.get(namespacedKey);
+            saveObjectToTags(namespacedKey, persistentDataObject, removeWhenEmpty);
+        }
+    }
+
+    void saveObjectToTags(@NotNull NamespacedKey key, @NotNull org.bukkit.persistence.PersistentDataObject persistentDataObject, boolean removeWhenEmpty){
+        try{
+            PersistentDataContainer serialized = persistentDataObject.serialize(parentContainer.getAdapterContext());
+            if(!serialized.isEmpty())
+                tags.put(key.toString(), registry.wrap(PersistentDataType.TAG_CONTAINER, PersistentDataType.TAG_CONTAINER.toPrimitive(serialized, parentContainer.getAdapterContext())));
+            else if(removeWhenEmpty) {
+                tags.remove(key.toString());
+            }
+            markParentAsDirty();
+        }
+        catch (Throwable e){
+            org.bukkit.Bukkit.getLogger().log(Level.WARNING, "An error occured while saving the persistent data object "+ persistentDataObject +" with the key "+key.asString()+" ", e);
+        }
+    }
+
+    void readObjectFromTags(@NotNull NamespacedKey key, @NotNull org.bukkit.persistence.PersistentDataObject persistentDataObject){
+        try{
+            if (parentContainer.has(key, PersistentDataType.TAG_CONTAINER)) {
+                PersistentDataContainer persistentDataContainer = parentContainer.get(key, PersistentDataType.TAG_CONTAINER);
+                if (persistentDataContainer != null)
+                    persistentDataObject.deSerialize(persistentDataContainer);
+            }
+        } catch (Throwable e){
+            org.bukkit.Bukkit.getLogger().log(Level.WARNING, "An error occured while loading the persistent data object "+ persistentDataObject +" with the key "+key.asString()+" ", e);
+        }
+    }
+
+    void serializeObjectAtKeyBeforeLookup(NamespacedKey key){
+        if(persistentDataObjectMap.containsKey(key))
+            saveObjectToTags(key, persistentDataObjectMap.get(key), true);
+    }
+
+    void markParentAsDirty(){
+        if(parentContainer instanceof DirtyCraftPersistentDataContainer dirtyCraftPersistentDataContainer)
+            dirtyCraftPersistentDataContainer.dirty(true);
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/persistence/DirtyCraftPersistentDataContainer.java b/src/main/java/org/bukkit/craftbukkit/persistence/DirtyCraftPersistentDataContainer.java
index fc3a34b59b70a6f553e28ba5c61d8ad50f3cb9b5..4ab6a291cf379b57410be292a7ff07e2d78285fe 100644
--- a/src/main/java/org/bukkit/craftbukkit/persistence/DirtyCraftPersistentDataContainer.java
+++ b/src/main/java/org/bukkit/craftbukkit/persistence/DirtyCraftPersistentDataContainer.java
@@ -60,4 +60,10 @@ public final class DirtyCraftPersistentDataContainer extends CraftPersistentData
         super.putAll(map);
         this.dirty(true);
     }
+
+    @Override
+    public CraftPersistentDataObjectCache getPersistentDataObjectCache() {
+        this.dirty(true);
+        return super.getPersistentDataObjectCache();
+    }
 }
