From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Thu, 21 Mar 2024 16:54:14 +0100
Subject: [PATCH] - Adding PredicateChoice


diff --git a/src/main/java/io/papermc/paper/inventory/recipe/RecipeBookExactChoiceRecipe.java b/src/main/java/io/papermc/paper/inventory/recipe/RecipeBookExactChoiceRecipe.java
index ef68600f6b59674ddea6c77f7e412902888e39b7..66bbf2471eaf475b982c4fead693df0a138481dc 100644
--- a/src/main/java/io/papermc/paper/inventory/recipe/RecipeBookExactChoiceRecipe.java
+++ b/src/main/java/io/papermc/paper/inventory/recipe/RecipeBookExactChoiceRecipe.java
@@ -6,25 +6,28 @@ import net.minecraft.world.item.crafting.Recipe;
 
 public abstract class RecipeBookExactChoiceRecipe<C extends net.minecraft.world.item.crafting.RecipeInput> implements Recipe<C> {
 
-    private boolean hasExactIngredients;
+    private boolean hasSpecialIngredients; // Paper - Adding PredicateChoice
 
     protected final void checkExactIngredients() {
         // skip any special recipes
         if (this.isSpecial()) {
-            this.hasExactIngredients = false;
+            this.hasSpecialIngredients = false; // Paper - Adding PredicateChoice
             return;
         }
         for (final Ingredient ingredient : this.getIngredients()) {
-            if (!ingredient.isEmpty() && ingredient.exact) {
-                this.hasExactIngredients = true;
+            // Paper start - Adding PredicateChoice
+            if (!ingredient.isEmpty() && (ingredient.exact || ingredient.itemPredicate != null)) {
+                this.hasSpecialIngredients = true;
+            // Paper end - Adding PredicateChoice
                 return;
             }
         }
-        this.hasExactIngredients = false;
+        this.hasSpecialIngredients = false; // Paper - Adding PredicateChoice
     }
 
     @Override
-    public final boolean hasExactIngredients() {
-        return this.hasExactIngredients;
-    }
+    // Paper start - Adding PredicateChoice
+    public final boolean hasSpecialIngredients() {
+        return this.hasSpecialIngredients;
+    } // Paper end - Adding PredicateChoice
 }
diff --git a/src/main/java/io/papermc/paper/inventory/recipe/StackedContentsExtraMap.java b/src/main/java/io/papermc/paper/inventory/recipe/StackedContentsExtraMap.java
index 568ba6aed2e74b8d84f4e82c1e785ef1587e2617..70a6b47f7cd5909209c763034c2fb48f2e0c17ab 100644
--- a/src/main/java/io/papermc/paper/inventory/recipe/StackedContentsExtraMap.java
+++ b/src/main/java/io/papermc/paper/inventory/recipe/StackedContentsExtraMap.java
@@ -28,6 +28,7 @@ public final class StackedContentsExtraMap {
     private final Int2ObjectMap<ItemStack> idToExactChoice = new Int2ObjectOpenHashMap<>();
     private final StackedContents contents;
     public final Map<Ingredient, IntList> extraStackingIds = new IdentityHashMap<>();
+    public final java.util.List<Ingredient> predicateChoices = new java.util.ArrayList<>();  // Paper - Adding PredicateChoice
 
     public StackedContentsExtraMap(final StackedContents contents, final Recipe<?> recipe) {
         this.exactChoiceIds.defaultReturnValue(-1);
@@ -36,7 +37,7 @@ public final class StackedContentsExtraMap {
     }
 
     private void initialize(final Recipe<?> recipe) {
-        if (recipe.hasExactIngredients()) {
+        if (recipe.hasSpecialIngredients()) { // Paper - Adding PredicateChoice
             for (final Ingredient ingredient : recipe.getIngredients()) {
                 if (!ingredient.isEmpty() && ingredient.exact) {
                     final net.minecraft.world.item.ItemStack[] items = ingredient.getItems();
@@ -51,6 +52,12 @@ public final class StackedContentsExtraMap {
                     idList.sort(IntComparators.NATURAL_COMPARATOR);
                     this.extraStackingIds.put(ingredient, idList);
                 }
+                // Paper start - Adding PredicateChoice
+                else if (!ingredient.isEmpty() && ingredient.itemPredicate != null) {
+                    this.predicateChoices.add(ingredient);
+                    this.extraStackingIds.put(ingredient, new IntArrayList()); // fill id list when accounting stacks
+                }
+                // Paper end - Adding PredicateChoice
             }
         }
     }
@@ -97,6 +104,18 @@ public final class StackedContentsExtraMap {
     }
 
     public boolean accountStack(final ItemStack stack, final int count) {
+        // Paper start - Adding PredicateChoice
+        // We are adding items that pass the predicate test.
+        for (final Ingredient predicateChoice : this.predicateChoices) {
+            if (predicateChoice.itemPredicate != null && predicateChoice.itemPredicate.test(stack.getBukkitStack())) {
+                final int id = this.registerExact(stack);
+                // We only want to add the stacking id to the list one time
+                if (id != -1) {
+                    this.extraStackingIds.get(predicateChoice).add(id);
+                }
+            }
+        }
+        // Paper end - Adding PredicateChoice
         if (!this.exactChoiceIds.isEmpty()) {
             final int id = this.exactChoiceIds.getInt(stack);
             if (id >= 0) {
diff --git a/src/main/java/net/minecraft/world/item/crafting/Ingredient.java b/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
index e314f36951e9ac15c57137e24fce8c410373130a..9fdb0a5373d21c63be414d1b49623015898dc0ea 100644
--- a/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
+++ b/src/main/java/net/minecraft/world/item/crafting/Ingredient.java
@@ -41,9 +41,9 @@ public final class Ingredient implements Predicate<ItemStack> {
     @Nullable
     private IntList stackingIds;
     public boolean exact; // CraftBukkit
+    @Nullable public Predicate<org.bukkit.inventory.ItemStack> itemPredicate; // Paper - Adding PredicateChoice
     public static final Codec<Ingredient> CODEC = Ingredient.codec(true);
     public static final Codec<Ingredient> CODEC_NONEMPTY = Ingredient.codec(false);
-
     public Ingredient(Stream<? extends Ingredient.Value> entries) {
         this.values = (Ingredient.Value[]) entries.toArray((i) -> {
             return new Ingredient.Value[i];
@@ -72,6 +72,11 @@ public final class Ingredient implements Predicate<ItemStack> {
         } else if (this.isEmpty()) {
             return itemstack.isEmpty();
         } else {
+            // Paper start - Adding PredicateChoice
+            if (itemPredicate != null) {
+                return itemPredicate.test(itemstack.getBukkitStack());
+            }
+            // Paper end - Adding PredicateChoice
             ItemStack[] aitemstack = this.getItems();
             int i = aitemstack.length;
 
diff --git a/src/main/java/net/minecraft/world/item/crafting/Recipe.java b/src/main/java/net/minecraft/world/item/crafting/Recipe.java
index 7fafbeacc80dfdeb0da40cf670dfe4c76104e7d3..2c4f13c73341623d73f5ef5bc543d85082bfd9e8 100644
--- a/src/main/java/net/minecraft/world/item/crafting/Recipe.java
+++ b/src/main/java/net/minecraft/world/item/crafting/Recipe.java
@@ -75,7 +75,7 @@ public interface Recipe<T extends RecipeInput> {
     org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id); // CraftBukkit
 
     // Paper start - improved exact choice recipes
-    default boolean hasExactIngredients() {
+    default boolean hasSpecialIngredients() { // Paper start - Adding PredicateChoice
         return false;
     }
     // Paper end
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
index c76c78bb7757d407102271463e14716a1b012deb..20d4c39bc39e72edb7002b901fddb1b778930de0 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
@@ -33,7 +33,15 @@ public interface CraftRecipe extends Recipe {
         } else if (bukkit == RecipeChoice.empty()) {
             stack = Ingredient.EMPTY;
             // Paper end
-        } else {
+        }
+        // Paper start - Adding PredicateChoice
+        else if(bukkit instanceof RecipeChoice.PredicateChoice predicateChoice){
+            List<org.bukkit.inventory.ItemStack> bukkitChoices = predicateChoice.recipeBookExamples();
+            stack = new Ingredient(bukkitChoices.stream().map(CraftItemStack::asNMSCopy).map(Ingredient.ItemValue::new));
+            stack.itemPredicate = predicateChoice.predicate();
+        }
+        // Paper end - Adding PredicateChoice
+        else {
             throw new IllegalArgumentException("Unknown recipe stack instance " + bukkit);
         }
 
@@ -51,7 +59,15 @@ public interface CraftRecipe extends Recipe {
         if (list.itemStacks.length == 0) {
             return RecipeChoice.empty(); // Paper - null breaks API contracts
         }
-
+        // Paper start - Adding PredicateChoice
+        if(list.itemPredicate != null) {
+            List<org.bukkit.inventory.ItemStack> choices = new ArrayList<>(list.itemStacks.length);
+            for (net.minecraft.world.item.ItemStack i : list.itemStacks) {
+                choices.add(CraftItemStack.asBukkitCopy(i));
+            }
+            return RecipeChoice.PredicateChoice.create(list.itemPredicate, choices);
+        }
+        // Paper end - Adding PredicateChoice
         if (list.exact) {
             List<org.bukkit.inventory.ItemStack> choices = new ArrayList<>(list.itemStacks.length);
             for (net.minecraft.world.item.ItemStack i : list.itemStacks) {
