From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Tue, 5 Dec 2023 21:21:01 +0100
Subject: [PATCH] Adding MCCreativeLabExtension plugin


diff --git a/build.gradle.kts b/build.gradle.kts
index 0dbf1035e26541d7d9e61329a04a6cb39e90b92b..ac1361ec9e19a5eebb2953cb02211e31aa0df263 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -201,12 +201,14 @@ tasks.registerRunTask("runShadow") {
     description = "Spin up a test server from the shadowJar archiveFile"
     classpath(tasks.shadowJar.flatMap { it.archiveFile })
     classpath(runtimeClasspathWithoutVanillaServer)
+    dependsOn(":plugin-extension:build")
 }
 
 tasks.registerRunTask("runReobf") {
     description = "Spin up a test server from the reobfJar output jar"
     classpath(tasks.reobfJar.flatMap { it.outputJar })
     classpath(runtimeClasspathWithoutVanillaServer)
+    dependsOn(":plugin-extension:build")
 }
 
 val runtimeClasspathForRunDev = sourceSets.main.flatMap { src ->
@@ -219,4 +221,5 @@ tasks.registerRunTask("runDev") {
     classpath(tasks.filterProjectDir.flatMap { it.outputJar })
     classpath(runtimeClasspathForRunDev)
     jvmArgs("-DPaper.isRunDev=true")
+    dependsOn(":plugin-extension:build")
 }
