From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Fri, 12 Jan 2024 16:01:12 +0100
Subject: [PATCH] Adding MilkGivingEntity to API


diff --git a/src/main/java/net/minecraft/world/entity/animal/Cow.java b/src/main/java/net/minecraft/world/entity/animal/Cow.java
index 3cdd9f379c7e2d46ea47c9ef55b121c93ec0bb4a..e4cb0219f988ed4af7a21818b1ee2f74f0fc5853 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cow.java
@@ -45,6 +45,7 @@ public class Cow extends Animal {
         super(type, world);
     }
 
+    public boolean canGiveMilk = true; // Paper - Can give milk API attribute
     @Override
     protected void registerGoals() {
         this.goalSelector.addGoal(0, new FloatGoal(this));
@@ -90,7 +91,13 @@ public class Cow extends Animal {
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemstack = player.getItemInHand(hand);
 
-        if (itemstack.is(Items.BUCKET) && !this.isBaby()) {
+        if (canGiveMilk && itemstack.is(Items.BUCKET) && !this.isBaby()) {
+            // Paper start - Can give milk API attribute
+            if(!canGiveMilk){
+                player.containerMenu.sendAllDataToRemote();
+                return super.mobInteract(player, hand);
+            }
+            // Paper end - Can give milk API attribute
             // CraftBukkit start - Got milk?
             PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent((ServerLevel) player.level(), player, this.blockPosition(), this.blockPosition(), null, itemstack, Items.MILK_BUCKET, hand);
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java b/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
index 111a244087e24f25ba8524a46a228da10cd9498a..a3263b8a3338edbe88fffd690b7d6caae0736aff 100644
--- a/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/src/main/java/net/minecraft/world/entity/animal/goat/Goat.java
@@ -76,6 +76,7 @@ public class Goat extends Animal {
     public static final EntityDataAccessor<Boolean> DATA_HAS_RIGHT_HORN = SynchedEntityData.defineId(Goat.class, EntityDataSerializers.BOOLEAN);
     private boolean isLoweringHead;
     private int lowerHeadTick;
+    public boolean canGiveMilk = true; // Paper - Can give milk API attribute
 
     public Goat(EntityType<? extends Goat> type, Level world) {
         super(type, world);
@@ -226,6 +227,12 @@ public class Goat extends Animal {
         ItemStack itemstack = player.getItemInHand(hand);
 
         if (itemstack.is(Items.BUCKET) && !this.isBaby()) {
+            // Paper start - Can give milk API attribute
+            if(!canGiveMilk){
+                player.containerMenu.sendAllDataToRemote();
+                return super.mobInteract(player, hand);
+            }
+            // Paper end - Can give milk API attribute
             // CraftBukkit start - Got milk?
             PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent((ServerLevel) player.level(), player, this.blockPosition(), this.blockPosition(), null, itemstack, Items.MILK_BUCKET, hand);
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftCow.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftCow.java
index 7babc404e4920cd264206d4e83b1be6f841cdb8c..a30124fe8b18af4811df3e8209f32f19dc32c3a8 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftCow.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftCow.java
@@ -18,4 +18,15 @@ public class CraftCow extends CraftAnimals implements Cow {
     public String toString() {
         return "CraftCow";
     }
+    // Paper start - // Paper - Can give milk API attribute
+    @Override
+    public boolean canGiveMilk() {
+        return getHandle().canGiveMilk;
+    }
+
+    @Override
+    public void setCanGiveMilk(boolean giveMilk) {
+        getHandle().canGiveMilk = giveMilk;
+    }
+    // Paper end - // Paper - Can give milk API attribute
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftGoat.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftGoat.java
index 2c21de478bff9cdf13ba46cd041831d54c11e924..5069d0333d088da35b44e2f5ee6c47a016e1cf0e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftGoat.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftGoat.java
@@ -55,4 +55,15 @@ public class CraftGoat extends CraftAnimals implements Goat {
         this.getHandle().ram(((CraftLivingEntity) entity).getHandle());
     }
     // Paper end
+    // Paper start - // Paper - Can give milk API attribute
+    @Override
+    public boolean canGiveMilk() {
+        return getHandle().canGiveMilk;
+    }
+
+    @Override
+    public void setCanGiveMilk(boolean giveMilk) {
+        getHandle().canGiveMilk = giveMilk;
+    }
+    // Paper end - // Paper - Can give milk API attribute
 }
