From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Wed, 29 May 2024 15:10:56 +0000
Subject: [PATCH] Add closing Inventory to InventoryCloseEvent if Reason =
 OPEN_NEW


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 7796e191747be545e744564a2b0b65790f69114d..49400ab3f977e3dad1413aabe7b71f4bb17ff436 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2861,6 +2861,16 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
 
     }
 
+    public void handleOpenContainerSwitch(ServerboundContainerClosePacket packet, InventoryView newView) {
+        // Paper end - Inventory close reason
+        PacketUtils.ensureRunningOnSameThread(packet, this, this.player.serverLevel());
+
+        if (this.player.isImmobile()) return; // CraftBukkit
+        CraftEventFactory.handleInventoryCloseEvent(this.player, org.bukkit.event.inventory.InventoryCloseEvent.Reason.OPEN_NEW, newView); // CraftBukkit // Paper
+
+        this.player.doCloseContainer();
+    }
+
     @Override
     public void handleContainerClose(ServerboundContainerClosePacket packet) {
         // Paper start - Inventory close reason
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index ba85b7bc8406c86b4f17b3aeab282598b5aecc43..3d263e5988f43d017da5fac6f50801c224f62b37 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -493,7 +493,7 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         if (((ServerPlayer) this.getHandle()).connection == null) return;
         if (this.getHandle().containerMenu != this.getHandle().inventoryMenu) {
             // fire INVENTORY_CLOSE if one already open
-            ((ServerPlayer) this.getHandle()).connection.handleContainerClose(new ServerboundContainerClosePacket(this.getHandle().containerMenu.containerId), org.bukkit.event.inventory.InventoryCloseEvent.Reason.OPEN_NEW); // Paper - Inventory close reason
+            ((ServerPlayer) this.getHandle()).connection.handleOpenContainerSwitch(new ServerboundContainerClosePacket(this.getHandle().containerMenu.containerId), inventory); // Paper - Inventory close reason
         }
         ServerPlayer player = (ServerPlayer) this.getHandle();
         AbstractContainerMenu container;
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 0ab53d46f0b8f3f3791dd01766738522c86932e8..9f45572ca0a876cdc9bd7837386d46abcffa4b0f 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1421,7 +1421,7 @@ public class CraftEventFactory {
     public static com.mojang.datafixers.util.Pair<net.kyori.adventure.text.@org.jetbrains.annotations.Nullable Component, @org.jetbrains.annotations.Nullable AbstractContainerMenu> callInventoryOpenEventWithTitle(ServerPlayer player, AbstractContainerMenu container, boolean cancelled) {
         // Paper end - Add titleOverride to InventoryOpenEvent
         if (player.containerMenu != player.inventoryMenu) { // fire INVENTORY_CLOSE if one already open
-            player.connection.handleContainerClose(new ServerboundContainerClosePacket(player.containerMenu.containerId), InventoryCloseEvent.Reason.OPEN_NEW); // Paper - Inventory close reason
+            player.connection.handleOpenContainerSwitch(new ServerboundContainerClosePacket(player.containerMenu.containerId), player.inventoryMenu.getBukkitView()); // Paper - Inventory close reason
         }
 
         CraftServer server = player.level().getCraftServer();
@@ -1616,6 +1616,15 @@ public class CraftEventFactory {
         return event;
     }
 
+    // MCCreativeLab start - Add Closing Inventory to InventoryCloseEvent when closed by another Inventory
+    public static void handleInventoryCloseEvent(net.minecraft.world.entity.player.Player human, org.bukkit.event.inventory.InventoryCloseEvent.Reason reason, @org.jetbrains.annotations.Nullable InventoryView newInventory) {
+        // Paper end
+        InventoryCloseEvent event = new InventoryCloseEvent(human.containerMenu.getBukkitView(), reason); // Paper
+        human.level().getCraftServer().getPluginManager().callEvent(event);
+        human.containerMenu.transferTo(human.inventoryMenu, human.getBukkitEntity());
+    }
+    // MCCreativeLab end - Add Closing Inventory to InventoryCloseEvent when closed by another Inventory
+
     // Paper start
     /**
      * Incase plugins hooked into this or Spigot adds a new inventory close event. Prefer to pass a reason
@@ -1623,13 +1632,10 @@ public class CraftEventFactory {
      */
     @Deprecated
     public static void handleInventoryCloseEvent(net.minecraft.world.entity.player.Player human) {
-        handleInventoryCloseEvent(human, org.bukkit.event.inventory.InventoryCloseEvent.Reason.UNKNOWN);
+        handleInventoryCloseEvent(human, org.bukkit.event.inventory.InventoryCloseEvent.Reason.UNKNOWN, null);
     }
     public static void handleInventoryCloseEvent(net.minecraft.world.entity.player.Player human, org.bukkit.event.inventory.InventoryCloseEvent.Reason reason) {
-        // Paper end
-        InventoryCloseEvent event = new InventoryCloseEvent(human.containerMenu.getBukkitView(), reason); // Paper
-        human.level().getCraftServer().getPluginManager().callEvent(event);
-        human.containerMenu.transferTo(human.inventoryMenu, human.getBukkitEntity());
+        handleInventoryCloseEvent(human, reason, null);
     }
 
     public static ItemStack handleEditBookEvent(ServerPlayer player, int itemInHandIndex, ItemStack itemInHand, ItemStack newBookItem) {
