From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Fri, 26 Jan 2024 15:15:53 +0100
Subject: [PATCH] Adds a new function to EntityBrain


diff --git a/src/main/java/de/verdox/mccreativelab/ai/CraftEntityBrain.java b/src/main/java/de/verdox/mccreativelab/ai/CraftEntityBrain.java
index 09b102bec6725b602d791c89ddef637797969207..1104c01be4cbaf41f16678f025769bad0487862c 100644
--- a/src/main/java/de/verdox/mccreativelab/ai/CraftEntityBrain.java
+++ b/src/main/java/de/verdox/mccreativelab/ai/CraftEntityBrain.java
@@ -165,6 +165,11 @@ public class CraftEntityBrain<E extends LivingEntity> implements EntityBrain<E>
         this.entityBrain.tick(((CraftWorld) world).getHandle(), ((CraftLivingEntity) entity).getHandle());
     }
 
+    @Override
+    public <U> void addMemoryToBrain(MemoryKey<U> memoryKey) {
+        this.entityBrain.addMemoryToBrain(CraftMemoryKey.bukkitToMinecraft(memoryKey));
+    }
+
     public static net.minecraft.world.entity.ai.memory.MemoryStatus toNMS(MemoryStatus memoryStatus) {
         return switch (memoryStatus) {
             case VALUE_PRESENT -> net.minecraft.world.entity.ai.memory.MemoryStatus.VALUE_PRESENT;
diff --git a/src/main/java/net/minecraft/world/entity/ai/Brain.java b/src/main/java/net/minecraft/world/entity/ai/Brain.java
index 8c7c4c99372d04ae22dfc2b8a3d62ce24da3ecca..be2c886a2609190afe1aa2f6d39d368e026a5f4b 100644
--- a/src/main/java/net/minecraft/world/entity/ai/Brain.java
+++ b/src/main/java/net/minecraft/world/entity/ai/Brain.java
@@ -409,6 +409,10 @@ public class Brain<E extends LivingEntity> {
     public Map<Integer, Map<Activity, Set<BehaviorControl<? super E>>>> getAvailableBehaviorsByPriority() {
         return availableBehaviorsByPriority;
     }
+
+    public void addMemoryToBrain(MemoryModuleType<?> memoryModuleType){
+        this.memories.put(memoryModuleType, Optional.empty());
+    }
     // MCCLab stop - Add CraftEntityBrain
 
     @VisibleForTesting
