From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Tue, 7 May 2024 22:38:09 +0000
Subject: [PATCH] Fixing randomTick behaviour method


diff --git a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
index 6d24bf837e28a1468f6033dd811d2c7e3dba44c4..38b9ed336c44b04cd7b191647b89038cdd370193 100644
--- a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -1346,9 +1346,7 @@ public abstract class BlockBehaviour implements FeatureElement {
 
         // MCCLab start - Add BlockBehaviour
         public final boolean isRandomlyTicking(Level world, int x, int y, int z, BlockState blockState){
-            if(world == null || x == -1 || z == -1 || blockState == null)
-                return this.isRandomlyTicking();
-            return blockBehaviour().isRandomlyTicking(world, x, y,z,blockState, this.isRandomlyTicking());
+            return blockBehaviour().isRandomlyTicking(world, x, y, z,blockState, this.isRandomlyTicking());
         }
 
         // MCCLab end - Add BlockBehaviour
diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
index 8d24184756b2c3406eb9e680f4e806dccbc1ebda..dc2e8de6914bedda628b1755d9c906fc74a891c0 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -12,6 +12,7 @@ import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
+import org.jetbrains.annotations.NotNull;
 
 public class LevelChunkSection {
 
@@ -29,7 +30,7 @@ public class LevelChunkSection {
     // Paper start - optimise collisions
     private int specialCollidingBlocks;
     // MCCLab start - Add BlockBehaviour
-    private net.minecraft.world.level.Level level;
+    @NotNull private final net.minecraft.world.level.Level level;
     private net.minecraft.world.level.ChunkPos chunkPos;
     private int chunkSectionY;
     // MCCLab end - Add BlockBehaviour
@@ -48,8 +49,10 @@ public class LevelChunkSection {
     }
     // Paper end - optimise collisions
 
-    public LevelChunkSection(PalettedContainer<BlockState> datapaletteblock, PalettedContainer<Holder<Biome>> palettedcontainerro) {
+    public LevelChunkSection(net.minecraft.world.level.Level level, net.minecraft.world.level.ChunkPos chunkPos, PalettedContainer<BlockState> datapaletteblock, PalettedContainer<Holder<Biome>> palettedcontainerro) {
         // CraftBukkit end
+        this.level = level;
+        this.chunkPos = chunkPos;
         this.states = datapaletteblock;
         this.biomes = palettedcontainerro;
         this.recalcBlockCounts();
diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/ChunkSerializer.java b/src/main/java/net/minecraft/world/level/chunk/storage/ChunkSerializer.java
index f3f2b5304595203923b288eaf60cb3a0c783386a..55bf8cf64e0f3491f42ce8a7fc546c811f56546a 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/ChunkSerializer.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/ChunkSerializer.java
@@ -207,7 +207,7 @@ public class ChunkSerializer {
                     object = new PalettedContainer<>(iregistry.asHolderIdMap(), iregistry.getHolderOrThrow(Biomes.PLAINS), PalettedContainer.Strategy.SECTION_BIOMES, null); // Paper - Anti-Xray - Add preset biomes
                 }
 
-                LevelChunkSection chunksection = new LevelChunkSection(datapaletteblock, (PalettedContainer) object); // CraftBukkit - read/write
+                LevelChunkSection chunksection = new LevelChunkSection(world, chunkPos, datapaletteblock, (PalettedContainer) object); // CraftBukkit - read/write
 
                 achunksection[k] = chunksection;
                 SectionPos sectionposition = SectionPos.of(chunkPos, b0);
