From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Fri, 26 Jan 2024 15:29:22 +0100
Subject: [PATCH] Small fix in CraftGoalFactory


diff --git a/src/main/java/de/verdox/mccreativelab/ai/builder/CraftGoalFactory.java b/src/main/java/de/verdox/mccreativelab/ai/builder/CraftGoalFactory.java
index 468f39572b8c191f3b765977425ccefa8ab70e1d..a1c86f6c3786fc53f5c98f7017c097c789eddfb1 100644
--- a/src/main/java/de/verdox/mccreativelab/ai/builder/CraftGoalFactory.java
+++ b/src/main/java/de/verdox/mccreativelab/ai/builder/CraftGoalFactory.java
@@ -14,6 +14,7 @@ import net.minecraft.world.entity.animal.Animal;
 import net.minecraft.world.entity.animal.ShoulderRidingEntity;
 import net.minecraft.world.entity.monster.CrossbowAttackMob;
 import net.minecraft.world.entity.monster.RangedAttackMob;
+import net.minecraft.world.entity.player.Player;
 import org.bukkit.Difficulty;
 import org.bukkit.Material;
 import org.bukkit.World;
@@ -293,7 +294,10 @@ public class CraftGoalFactory implements GoalFactory {
         return asGoal(new ZombieAttackGoal((net.minecraft.world.entity.monster.Zombie) toNms(zombie), speed, pauseWhenMobIdle));
     }
 
-    private Class<? extends net.minecraft.world.entity.PathfinderMob> toNms(World world, EntityType entityType) {
+    private Class<? extends net.minecraft.world.entity.LivingEntity> toNms(World world, EntityType entityType) {
+        if(entityType.equals(EntityType.PLAYER))
+            return Player.class;
+
         CraftRegionAccessor craftRegionAccessor = (CraftRegionAccessor) world;
         return (Class<? extends PathfinderMob>) ((CraftEntity) craftRegionAccessor.createEntity(world.getSpawnLocation(), entityType.getEntityClass()))
             .getHandle().getClass();
