From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Wed, 19 Jun 2024 15:13:05 +0000
Subject: [PATCH] Flipped Custom Coordinate Flag in CraftrAdvancementBuilder


diff --git a/src/main/java/de/verdox/mccreativelab/advancement/CraftAdvancementBuilder.java b/src/main/java/de/verdox/mccreativelab/advancement/CraftAdvancementBuilder.java
index c5aa73f5c34dfd49f913bdc7e053f3cbe376fb29..e9f7d855483fe0a096f1303aca77a7e0fa978105 100644
--- a/src/main/java/de/verdox/mccreativelab/advancement/CraftAdvancementBuilder.java
+++ b/src/main/java/de/verdox/mccreativelab/advancement/CraftAdvancementBuilder.java
@@ -84,7 +84,7 @@ public class CraftAdvancementBuilder implements AdvancementBuilder {
         final net.minecraft.advancements.AdvancementTree tree = MinecraftServer.getServer().getAdvancements().tree();
         tree.addAll(List.of(holder));
 
-        if (craftAdvancementDisplayBuilder.customCoordinates) {
+        if (!craftAdvancementDisplayBuilder.customCoordinates) {
             final net.minecraft.advancements.AdvancementNode node = tree.get(minecraftkey);
             if (node != null) {
                 final net.minecraft.advancements.AdvancementNode root = node.root();
diff --git a/src/main/java/de/verdox/mccreativelab/advancement/CraftAdvancementDisplayBuilder.java b/src/main/java/de/verdox/mccreativelab/advancement/CraftAdvancementDisplayBuilder.java
index d3f6890d1f2da9a790ab14cae299958bdd413ef4..b375d753666dff8253b849ffe9736d27c987dee5 100644
--- a/src/main/java/de/verdox/mccreativelab/advancement/CraftAdvancementDisplayBuilder.java
+++ b/src/main/java/de/verdox/mccreativelab/advancement/CraftAdvancementDisplayBuilder.java
@@ -81,12 +81,14 @@ public class CraftAdvancementDisplayBuilder implements AdvancementDisplayBuilder
     @Override
     public AdvancementDisplayBuilder setX(float x) {
         this.x = x;
+        this.customCoordinates = true;
         return this;
     }
 
     @Override
     public AdvancementDisplayBuilder setY(float y) {
         this.y = y;
+        this.customCoordinates = true;
         return this;
     }
 
