From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Thu, 22 Aug 2024 21:18:44 +0000
Subject: [PATCH] Don't throw exception when calling setSpectatorTarget when
 not in spectator mode


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 918cb8c84339682260f14ab1ff8da88bf9b0d2ac..f73f510b2e3ea93d7987764533d5808ac8d175f7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2918,7 +2918,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public void setSpectatorTarget(org.bukkit.entity.Entity entity) {
-        Preconditions.checkArgument(this.getGameMode() == GameMode.SPECTATOR, "Player must be in spectator mode");
+        //Preconditions.checkArgument(this.getGameMode() == GameMode.SPECTATOR, "Player must be in spectator mode");
         this.getHandle().setCamera((entity == null) ? null : ((CraftEntity) entity).getHandle());
     }
 
