From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Wed, 15 Nov 2023 17:47:40 +0100
Subject: [PATCH] Adding growTreeIfSapling helper method to Bukkit World class


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index d8a23aa0d898ca3360757721e38ddb97387f7d21..dbc0b8ed80d5dc1acb1bd4b822f879f234e3ad2d 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -52,6 +52,17 @@ import org.jetbrains.annotations.Nullable;
  */
 public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient, Metadatable, PersistentDataHolder, Keyed, net.kyori.adventure.audience.ForwardingAudience { // Paper
 
+    // MCCLab - start
+    /**
+     * Creates a tree at the given {@link Location} if the {@link BlockData} of the {@link Block} is of type {@link org.bukkit.block.data.type.Sapling}
+     *
+     * @param location Location to grow the tree
+     * @return true if the tree was created successfully, otherwise false
+     */
+    boolean growTreeIfSapling(@NotNull Location location);
+
+
+    // MCCLab - end
     // Paper start
     /**
      * @return The amount of Entities in this world
