From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Fri, 17 Nov 2023 17:05:04 +0100
Subject: [PATCH] Adding isRainingAt method to World.java


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index dbc0b8ed80d5dc1acb1bd4b822f879f234e3ad2d..613d8fc14c1a86c232691dc787ffe86f95e16182 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -53,6 +53,14 @@ import org.jetbrains.annotations.Nullable;
 public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient, Metadatable, PersistentDataHolder, Keyed, net.kyori.adventure.audience.ForwardingAudience { // Paper
 
     // MCCLab - start
+
+    /**
+     * Checks if it is raining at a specific location
+     * @param location The location to check
+     * @return true if it is raining at the block location
+     */
+    boolean isRainingAt(Location location);
+
     /**
      * Creates a tree at the given {@link Location} if the {@link BlockData} of the {@link Block} is of type {@link org.bukkit.block.data.type.Sapling}
      *
