From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sat, 20 Jan 2024 16:51:20 +0100
Subject: [PATCH] Ignoring errors when creating javadocs


diff --git a/build.gradle.kts b/build.gradle.kts
index 540fe7e2c110e79c3742f229b3ed8c54b101d260..709cac0c2a305ec1162c77550530a0d3afee7248 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -146,6 +146,7 @@ tasks.withType<Javadoc> {
     options.overview = "src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
+    isFailOnError = false;
     options.links(
         "https://guava.dev/releases/32.1.2-jre/api/docs/",
         "https://javadoc.io/doc/org.yaml/snakeyaml/2.2/",
