From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sat, 20 Jan 2024 16:51:20 +0100
Subject: [PATCH] Ignoring errors when creating javadocs


diff --git a/build.gradle.kts b/build.gradle.kts
index ab84a1405acc1f0d5f267892243b82b8dab03e21..124fe3a861fd77b5d4d3f92eff489f1229b818a2 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -147,6 +147,7 @@ tasks.withType<Javadoc> {
     options.overview = "src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
+    isFailOnError = false;
     options.links(
         "https://guava.dev/releases/32.1.2-jre/api/docs/",
         "https://javadoc.io/doc/org.yaml/snakeyaml/2.2/",
