From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Sat, 18 May 2024 23:18:52 +0000
Subject: [PATCH] Adding the ability to change the server resourcepack


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 08499fdacdabecf16adebd89cba0820d4431dd73..85735cf1228f4b95df4d308a2854a79f1a99945f 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -349,6 +349,16 @@ public final class Bukkit {
         return server.getServerResourcePack();
     }
 
+    // Paper start - Make server resourcepack changeable
+    /**
+     * Sets the server resource pack
+     * @param resourcePack the resource pack
+     */
+    public static void setServerResourcePack(@Nullable ResourcePack resourcePack){
+        server.setServerResourcePack(resourcePack);
+    }
+    // Paper end - Make server resourcepack changeable
+
     /**
      * Get the ServerTick Manager.
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 0892fab69105854c4f5a75f1f3796c6b6534a26a..68b45cc4552d3056f795bc940f3b87173c96a735 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -292,6 +292,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @Nullable
     public ResourcePack getServerResourcePack();
 
+    // Paper start - Make server resourcepack changeable
+    /**
+     * Sets the server resource pack
+     * @param resourcePack the resource pack
+     */
+    public void setServerResourcePack(@Nullable ResourcePack resourcePack);
+    // Paper end - Make server resourcepack changeable
+
     /**
      * Gets the server resource pack uri, or empty string if not specified.
      *
diff --git a/src/main/java/org/bukkit/packs/ResourcePack.java b/src/main/java/org/bukkit/packs/ResourcePack.java
index 2862fb5bab340f027fecacf7ccf0df50f26a91bf..7a286e49f5797b19319b482d1ab1188d4ef9fcce 100644
--- a/src/main/java/org/bukkit/packs/ResourcePack.java
+++ b/src/main/java/org/bukkit/packs/ResourcePack.java
@@ -41,7 +41,7 @@ public interface ResourcePack {
      * @return the prompt
      */
     @Nullable
-    public String getPrompt();
+    public net.kyori.adventure.text.Component getPrompt(); // Paper - Make server resourcepack changeable
 
     /**
      * Gets if the resource pack is required by the server.
diff --git a/src/test/java/org/bukkit/plugin/TimedRegisteredListenerTest.java b/src/test/java/org/bukkit/plugin/TimedRegisteredListenerTest.java
index 9ed416ed57676c845833736f93ed6088513c6da4..90ee4cdc682434aed66a98a687b2c549a6f381c4 100644
--- a/src/test/java/org/bukkit/plugin/TimedRegisteredListenerTest.java
+++ b/src/test/java/org/bukkit/plugin/TimedRegisteredListenerTest.java
@@ -25,7 +25,7 @@ public class TimedRegisteredListenerTest {
 
         PlayerInteractEvent interactEvent = new PlayerInteractEvent(null, null, null, null, null);
         PlayerMoveEvent moveEvent = new PlayerMoveEvent(null, null, null);
-        BlockBreakEvent breakEvent = new BlockBreakEvent(null, null);
+        BlockBreakEvent breakEvent = new BlockBreakEvent(null, null, null);
 
         TimedRegisteredListener trl = new TimedRegisteredListener(listener, executor, EventPriority.NORMAL, plugin, false);
 
