From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Thu, 27 Jun 2024 15:26:53 +0000
Subject: [PATCH] Fixes


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 1b5358c6ab078fba6d9d6fd80bf3ce5edb3a4d1d..5faabcb0888cab534c7ef63823add0f73072bef7 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -2787,7 +2787,7 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     int getSimulationDistance();
 
     /**
-     * Returns the {@link WorldGenData} of this world
+     * Returns the {@link io.papermc.paper.world.worldgen.WorldGenData} of this world
      * @return the world gen data
      */
     public io.papermc.paper.world.worldgen.WorldGenData getWorldGenData();
