From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Thu, 21 Mar 2024 17:57:27 +0100
Subject: [PATCH] Adding resendChunksForBiomes function


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 74cd2aeb535e70047f0096ac80183adee7456024..7fcfdf0be9b4a8d6c71bcd74de00cabb0898dc98 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -2345,6 +2345,12 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @Nullable
     public BiomeProvider getBiomeProvider();
 
+    /**
+     * Resends chunks to players
+     * @param chunks - The chunks to resend
+     */
+    public void resendChunksForBiomes(List<Chunk> chunks);
+
     /**
      * Saves world to disk
      */
