From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Wed, 15 Nov 2023 17:31:15 +0100
Subject: [PATCH] Adding helper methods for getting and setting customModelData
 on ItemStacks


diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index 0af73cc04edb93b9772136d4d808f657ea40e733..7494ab7cb0db72cc509b9890eb98777c356ae924 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -26,6 +26,34 @@ import org.jetbrains.annotations.Nullable;
  * returns false.</b>
  */
 public class ItemStack implements Cloneable, ConfigurationSerializable, Translatable, net.kyori.adventure.text.event.HoverEventSource<net.kyori.adventure.text.event.HoverEvent.ShowItem>, net.kyori.adventure.translation.Translatable { // Paper
+
+    // MCCLab - start
+    /**
+     * Helper method to set customModelData of an ItemStack
+     * @param stack the ItemStack
+     * @param customModelData the new customModelData
+     */
+    public static void setCustomModelData(ItemStack stack, int customModelData){
+        stack.editMeta(meta -> meta.setCustomModelData(customModelData));
+    }
+
+    /**
+     * Helper method to get the customModelData of an ItemStack
+     * @param stack the ItemStack
+     * @return the customModelData
+     */
+
+    public static int getCustomModelData(ItemStack stack){
+        if(stack == null || stack.getType().isAir())
+            return 0;
+        if(!stack.hasItemMeta())
+            return 0;
+        if(!stack.getItemMeta().hasCustomModelData())
+            return 0;
+        return stack.getItemMeta().getCustomModelData();
+    }
+    // MCCLab - end
+
     private Material type = Material.AIR;
     private int amount = 0;
     private MaterialData data = null;
